{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Constructing ISI's COVID-19 Knowledge Graph"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This notebook shows how we use our KGTK toolking (link) to construct the COVID-19 knowledge graph for the CORD-19 corpus using the text extractions from the BLENDER group at UIUC. The text extractions inlcude a variety of entity extractions with links to bioinformatics databases. Our approach is to:\n",
    "\n",
    "* extract from Wikidata the subgraph that covers all the publications in the CORD-19 corpus and the entities identified by the BLENDER group.\n",
    "* define new Wikidata items for publications or entities that are not present in the public Wikidata\n",
    "* annotate the article items with the relevant entities\n",
    "* preserve provenance\n",
    "\n",
    "We implement the approach in the following steps:\n",
    "\n",
    "* data prepararation: convert the JSON representation of BLENDER output to CSV files that are easy to process\n",
    "* extract Wikidata subgraph: extract from Wikidata the articles, authors, and entities mentioned in the BLENDER corpus\n",
    "* create missing items: create nodes for articles and entities that are not present in Wikidata\n",
    "* create mention edges: create edges to record the entity extractions from BLENDER, including justifications\n",
    "* incorporate analytic outputs: add edges to record graph metrics such as pagerank\n",
    "* export knowledge graph: export the graph to KGTL edges, RDF and Neo4J "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data Preparation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Set up environment variables with location of the input files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "env: COVID=/Users/pedroszekely/data/covid/blender\n",
      "env: WD=/Volumes/GoogleDrive/Shared drives/KGTK/datasets/wikidata-20200504\n"
     ]
    }
   ],
   "source": [
    "%env COVID=/Users/pedroszekely/data/covid/blender\n",
    "%env WD=/Volumes/GoogleDrive/Shared drives/KGTK/datasets/wikidata-20200504"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Wikidata Data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The Wikidata files are the large:\n",
    "\n",
    "* `wikidata_nodes_20200504.tsv.gz` the English labels, alias and descriptions for all items in Wikidata\n",
    "* `wikidata_edges_20200504.tsv.gz` the edges for all statements in Wikidata\n",
    "* `wikidata_qualifiers_20200504.tsv.gz` the qualifier edges for all statements in Wikidata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 44833280\n",
      "-rw-------  1 pedroszekely  staff    21M May 11 14:19 P279.tsv.gz\n",
      "-rw-------  1 pedroszekely  staff   6.9M May 11 15:59 P279_star.tsv.gz\n",
      "-rw-------  1 pedroszekely  staff    12M May 11 14:29 P279_truncated.tsv.gz\n",
      "-rw-------  1 pedroszekely  staff   500M May 11 12:53 P31.tsv.gz\n",
      "-rw-------  1 pedroszekely  staff    16G May 11 04:09 wikidata_edges_20200504.tsv.gz\n",
      "-rw-------  1 pedroszekely  staff   2.2G May 11 03:22 wikidata_nodes_20200504.tsv.gz\n",
      "-rw-------  1 pedroszekely  staff   2.4G May 11 04:19 wikidata_qualifiers_20200504.tsv.gz\n"
     ]
    }
   ],
   "source": [
    "ls -lh \"$WD\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Working with the dump files takes time because they are so large (86B chars and 1.1B lines), more than 6 minutes to just unzip and count lines"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " 1105944516 7412374276 86582473531\n",
      "\n",
      "real\t6m29.270s\n",
      "user\t8m28.469s\n",
      "sys\t1m17.519s\n"
     ]
    }
   ],
   "source": [
    "!time gzcat \"$WD/wikidata_edges_20200504.tsv.gz\" | wc"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The columns in the edge files:\n",
    "\n",
    "* `id` is a unique identifier for an edge, and provides an identifier for each statement in Wikidata\n",
    "* `node1`, `label` and `node2` are the item/property/value or subject/predicate/object\n",
    "* `rank` is the Wikidata rank for the statement\n",
    "* `node2;*` are additional columns that provide detailed information about the `node2` column, making it easy to parse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "id\tnode1\tlabel\tnode2\trank\tnode2;magnitude\tnode2;unit\tnode2;date\tnode2;item\tnode2;lower\tnode2;upper\tnode2;latitude\tnode2;longitude\tnode2;precision\tnode2;calendar\tnode2;entity-type\n",
      "Q8-P1245-1\tQ8\tP1245\t\"885155\"\tnormal\t\t\t\t\t\t\t\t\t\t\t\n",
      "Q8-P373-1\tQ8\tP373\t\"Happiness\"\tnormal\t\t\t\t\t\t\t\t\t\t\t\n",
      "Q8-P31-1\tQ8\tP31\tQ331769\tnormal\t\t\t\tQ331769\t\t\t\t\t\t\titem\n",
      "Q8-P31-2\tQ8\tP31\tQ60539479\tnormal\t\t\t\tQ60539479\t\t\t\t\t\t\titem\n",
      "Q8-P31-3\tQ8\tP31\tQ9415\tnormal\t\t\t\tQ9415\t\t\t\t\t\t\titem\n",
      "Q8-P508-1\tQ8\tP508\t\"13163\"\tnormal\t\t\t\t\t\t\t\t\t\t\t\n",
      "Q8-P18-1\tQ8\tP18\t\"Sweet Baby Kisses Family Love.jpg\"\tnormal\t\t\t\t\t\t\t\t\t\t\t\n",
      "Q8-P910-1\tQ8\tP910\tQ8505256\tnormal\t\t\t\tQ8505256\t\t\t\t\t\t\titem\n",
      "Q8-P349-1\tQ8\tP349\t\"00566227\"\tnormal\t\t\t\t\t\t\t\t\t\t\t\n",
      "gzcat: error writing to output: Broken pipe\n",
      "gzcat: /Volumes/GoogleDrive/Shared drives/KGTK/datasets/wikidata-20200504/wikidata_edges_20200504.tsv.gz: uncompress failed\n"
     ]
    }
   ],
   "source": [
    "!gzcat \"$WD/wikidata_edges_20200504.tsv.gz\" | head"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### BLENDER Data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The BLENDER data came in a JSON document per article. We used Python scripts to create simple TSV files that are easy to process.\n",
    "\n",
    "The `corpus-identifiers.tsv` file contains all the identifiers present in the BLENDER dataset with two columns:\n",
    "\n",
    "* node2: the value of the identifier\n",
    "* label: the name of the property in Wikidata used to represent the specific identifier. For example `P698` is PubMed ID. See https://www.wikidata.org/wiki/Q93157077"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node2</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PMC3670673</td>\n",
       "      <td>P932</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3670673</td>\n",
       "      <td>P932</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>22621853</td>\n",
       "      <td>P698</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9606</td>\n",
       "      <td>P685</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>851819</td>\n",
       "      <td>P351</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139032</th>\n",
       "      <td>19423234</td>\n",
       "      <td>P698</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139033</th>\n",
       "      <td>22609285</td>\n",
       "      <td>P698</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139034</th>\n",
       "      <td>12781505</td>\n",
       "      <td>P698</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139035</th>\n",
       "      <td>10028170</td>\n",
       "      <td>P698</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139036</th>\n",
       "      <td>17961784</td>\n",
       "      <td>P698</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>139037 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             node2 label\n",
       "0       PMC3670673  P932\n",
       "1          3670673  P932\n",
       "2         22621853  P698\n",
       "3             9606  P685\n",
       "4           851819  P351\n",
       "...            ...   ...\n",
       "139032    19423234  P698\n",
       "139033    22609285  P698\n",
       "139034    12781505  P698\n",
       "139035    10028170  P698\n",
       "139036    17961784  P698\n",
       "\n",
       "[139037 rows x 2 columns]"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ids = pd.read_csv(os.getenv(\"COVID\")+'/corpus-identifiers.tsv', delimiter='\\t')\n",
    "ids"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extract Wikidata Subgraph"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Step 1: find the `node1` for all the rows in `corpus-identifiers.tsv`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We do this with the `kgtk ifexists` command to scan the file of all edges in Wikidata to select the ones where `label` and `node2` match the rows in `corpus-identifiers.tsv`. We store the results in `corpus-identifier-edges.tsv`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "real\t50m53.543s\n",
      "user\t52m18.007s\n",
      "sys\t2m9.164s\n"
     ]
    }
   ],
   "source": [
    "!time gzcat \"$WD/wikidata_edges_20200504.tsv.gz\" \\\n",
    "  | kgtk ifexists --input-keys label node2 --filter-on $COVID/corpus-identifiers.tsv --filter-keys label node2 --filter-mode NONE \\\n",
    "  > $COVID/corpus-identifier-edges.tsv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We found 72,649 q-nodes in Wikidata for the identifiers in our corpus. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>node1</th>\n",
       "      <th>label</th>\n",
       "      <th>node2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Q140-P685-1</td>\n",
       "      <td>Q140</td>\n",
       "      <td>P685</td>\n",
       "      <td>9689</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q556-P486-1</td>\n",
       "      <td>Q556</td>\n",
       "      <td>P486</td>\n",
       "      <td>D006859</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Q688-P486-1</td>\n",
       "      <td>Q688</td>\n",
       "      <td>P486</td>\n",
       "      <td>D002713</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Q716-P486-1</td>\n",
       "      <td>Q716</td>\n",
       "      <td>P486</td>\n",
       "      <td>D014025</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Q1832-P486-1</td>\n",
       "      <td>Q1832</td>\n",
       "      <td>P486</td>\n",
       "      <td>D005682</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72644</th>\n",
       "      <td>Q93147847-P932-1</td>\n",
       "      <td>Q93147847</td>\n",
       "      <td>P932</td>\n",
       "      <td>6730851</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72645</th>\n",
       "      <td>Q93157077-P698-1</td>\n",
       "      <td>Q93157077</td>\n",
       "      <td>P698</td>\n",
       "      <td>31492122</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72646</th>\n",
       "      <td>Q93157077-P932-1</td>\n",
       "      <td>Q93157077</td>\n",
       "      <td>P932</td>\n",
       "      <td>6731609</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72647</th>\n",
       "      <td>Q93157501-P698-1</td>\n",
       "      <td>Q93157501</td>\n",
       "      <td>P698</td>\n",
       "      <td>31492169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72648</th>\n",
       "      <td>Q93157501-P932-1</td>\n",
       "      <td>Q93157501</td>\n",
       "      <td>P932</td>\n",
       "      <td>6729025</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>72649 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     id      node1 label     node2\n",
       "0           Q140-P685-1       Q140  P685      9689\n",
       "1           Q556-P486-1       Q556  P486   D006859\n",
       "2           Q688-P486-1       Q688  P486   D002713\n",
       "3           Q716-P486-1       Q716  P486   D014025\n",
       "4          Q1832-P486-1      Q1832  P486   D005682\n",
       "...                 ...        ...   ...       ...\n",
       "72644  Q93147847-P932-1  Q93147847  P932   6730851\n",
       "72645  Q93157077-P698-1  Q93157077  P698  31492122\n",
       "72646  Q93157077-P932-1  Q93157077  P932   6731609\n",
       "72647  Q93157501-P698-1  Q93157501  P698  31492169\n",
       "72648  Q93157501-P932-1  Q93157501  P932   6729025\n",
       "\n",
       "[72649 rows x 4 columns]"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "idedges = pd.read_csv(os.getenv(\"COVID\")+'/corpus-identifier-edges.tsv', delimiter='\\t')\n",
    "idedges.loc[:, ['id', 'node1', 'label', 'node2']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here are the counts of the q-nodes we found in Wikidata for each property."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "P698     20426\n",
       "P351     18235\n",
       "P932     14169\n",
       "P486     10096\n",
       "P685      9721\n",
       "P5055        2\n",
       "Name: label, dtype: int64"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "idedges.label.value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Step 2: get all the edges from Wikidata for the q-nodes in `corpus-identifier-edges.tsv`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To do this, we again scan all the edges in Wikidata looking for edges whose `node1` matches the `node1` in `corpus-identifier-edges.tsv`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "real\t42m33.188s\n",
      "user\t44m19.812s\n",
      "sys\t1m50.926s\n"
     ]
    }
   ],
   "source": [
    "!time gzcat \"$WD/wikidata_edges_20200504.tsv.gz\" \\\n",
    "  | kgtk ifexists --input-keys node1 --filter-on $COVID/corpus-identifier-edges.tsv --filter-keys node1 \\\n",
    "  > $COVID/corpus-edges.tsv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We now have 2.2 million edges for the entities in our corpus."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " 2297783 13662891 176151115 /Users/pedroszekely/data/covid/blender/corpus-edges.tsv\n"
     ]
    }
   ],
   "source": [
    "!wc $COVID/corpus-edges.tsv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can use Pandas to explore the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/pedroszekely/opt/anaconda3/envs/kgtk-env/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3063: DtypeWarning: Columns (6,7,14) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>node1</th>\n",
       "      <th>label</th>\n",
       "      <th>node2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Q140-P225-1</td>\n",
       "      <td>Q140</td>\n",
       "      <td>P225</td>\n",
       "      <td>Panthera leo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q140-P105-1</td>\n",
       "      <td>Q140</td>\n",
       "      <td>P105</td>\n",
       "      <td>Q7432</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Q140-P171-1</td>\n",
       "      <td>Q140</td>\n",
       "      <td>P171</td>\n",
       "      <td>Q127960</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Q140-P31-1</td>\n",
       "      <td>Q140</td>\n",
       "      <td>P31</td>\n",
       "      <td>Q16521</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Q140-P1403-1</td>\n",
       "      <td>Q140</td>\n",
       "      <td>P1403</td>\n",
       "      <td>Q15294488</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2297777</th>\n",
       "      <td>Q93157501-P478-1</td>\n",
       "      <td>Q93157501</td>\n",
       "      <td>P478</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2297778</th>\n",
       "      <td>Q93157501-P50-1</td>\n",
       "      <td>Q93157501</td>\n",
       "      <td>P50</td>\n",
       "      <td>Q92401632</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2297779</th>\n",
       "      <td>Q93157501-P577-1</td>\n",
       "      <td>Q93157501</td>\n",
       "      <td>P577</td>\n",
       "      <td>^2019-09-06T00:00:00Z/11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2297780</th>\n",
       "      <td>Q93157501-P698-1</td>\n",
       "      <td>Q93157501</td>\n",
       "      <td>P698</td>\n",
       "      <td>31492169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2297781</th>\n",
       "      <td>Q93157501-P932-1</td>\n",
       "      <td>Q93157501</td>\n",
       "      <td>P932</td>\n",
       "      <td>6729025</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2297782 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       id      node1  label                     node2\n",
       "0             Q140-P225-1       Q140   P225              Panthera leo\n",
       "1             Q140-P105-1       Q140   P105                     Q7432\n",
       "2             Q140-P171-1       Q140   P171                   Q127960\n",
       "3              Q140-P31-1       Q140    P31                    Q16521\n",
       "4            Q140-P1403-1       Q140  P1403                 Q15294488\n",
       "...                   ...        ...    ...                       ...\n",
       "2297777  Q93157501-P478-1  Q93157501   P478                        17\n",
       "2297778   Q93157501-P50-1  Q93157501    P50                 Q92401632\n",
       "2297779  Q93157501-P577-1  Q93157501   P577  ^2019-09-06T00:00:00Z/11\n",
       "2297780  Q93157501-P698-1  Q93157501   P698                  31492169\n",
       "2297781  Q93157501-P932-1  Q93157501   P932                   6729025\n",
       "\n",
       "[2297782 rows x 4 columns]"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ce = pd.read_csv(os.getenv(\"COVID\")+'/corpus-edges.tsv', delimiter='\\t')\n",
    "ce.loc[:, ['id', 'node1', 'label', 'node2']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "P2860                 540720\n",
       "wikipedia_sitelink    300857\n",
       "P704                  120845\n",
       "P2093                 109421\n",
       "P639                   87783\n",
       "                       ...  \n",
       "P4611                      1\n",
       "P3100                      1\n",
       "P1268                      1\n",
       "P3252                      1\n",
       "P5420                      1\n",
       "Name: label, Length: 952, dtype: int64"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ce.label.value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can use the Wikidata cli to see what the top properties:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[90mid\u001b[39m P2860\n",
      "\u001b[42mLabel\u001b[49m cites work\n",
      "\u001b[44mDescription\u001b[49m citation from one creative work to another\n",
      "\u001b[30m\u001b[47minstance of\u001b[49m\u001b[39m \u001b[90m(P31)\u001b[39m\u001b[90m: \u001b[39mWikidata property for items about works \u001b[90m(Q18618644)\u001b[39m\n",
      "\n",
      "\u001b[90mid\u001b[39m P704\n",
      "\u001b[42mLabel\u001b[49m Ensembl transcript ID\n",
      "\u001b[44mDescription\u001b[49m transcript ID issued by Ensembl database\n",
      "\u001b[30m\u001b[47minstance of\u001b[49m\u001b[39m \u001b[90m(P31)\u001b[39m\u001b[90m: \u001b[39mWikidata property for an identifier \u001b[90m(Q19847637)\u001b[39m | Wikidata property related to medicine \u001b[90m(Q19887775)\u001b[39m\n",
      "\n",
      "\u001b[90mid\u001b[39m P2093\n",
      "\u001b[42mLabel\u001b[49m author name string\n",
      "\u001b[44mDescription\u001b[49m string to store unspecified author name for publications; use if Wikidata item for author (P50) does not exist or is not known\n",
      "\u001b[30m\u001b[47minstance of\u001b[49m\u001b[39m \u001b[90m(P31)\u001b[39m\u001b[90m: \u001b[39mWikidata property for items about works \u001b[90m(Q18618644)\u001b[39m | Wikidata property with datatype string that is not an external identifier \u001b[90m(Q21099935)\u001b[39m | Wikidata property to indicate a source \u001b[90m(Q18608359)\u001b[39m\n"
     ]
    }
   ],
   "source": [
    "!wd u P2860, P704, P2093"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "How many authors have items in Wikidata?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   25284\n"
     ]
    }
   ],
   "source": [
    "!kgtk filter -p ';P50;' $COVID/corpus-edges.tsv | wc -l"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Step 3:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "!kgtk filter -p ';P50, P2860;' $COVID/corpus-edges.tsv > $COVID/citation-and-author-edges.tsv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>node1</th>\n",
       "      <th>label</th>\n",
       "      <th>node2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Q21090495-P2860-1</td>\n",
       "      <td>Q21090495</td>\n",
       "      <td>P2860</td>\n",
       "      <td>Q24611162</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q21090495-P2860-2</td>\n",
       "      <td>Q21090495</td>\n",
       "      <td>P2860</td>\n",
       "      <td>Q24655519</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Q21090495-P2860-3</td>\n",
       "      <td>Q21090495</td>\n",
       "      <td>P2860</td>\n",
       "      <td>Q22065976</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Q21090495-P2860-4</td>\n",
       "      <td>Q21090495</td>\n",
       "      <td>P2860</td>\n",
       "      <td>Q24650035</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Q21090495-P2860-5</td>\n",
       "      <td>Q21090495</td>\n",
       "      <td>P2860</td>\n",
       "      <td>Q24684593</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>565998</th>\n",
       "      <td>Q93147847-P50-1</td>\n",
       "      <td>Q93147847</td>\n",
       "      <td>P50</td>\n",
       "      <td>Q61104970</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>565999</th>\n",
       "      <td>Q93147847-P50-2</td>\n",
       "      <td>Q93147847</td>\n",
       "      <td>P50</td>\n",
       "      <td>Q90414144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>566000</th>\n",
       "      <td>Q93147847-P50-3</td>\n",
       "      <td>Q93147847</td>\n",
       "      <td>P50</td>\n",
       "      <td>Q87706998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>566001</th>\n",
       "      <td>Q93157077-P50-1</td>\n",
       "      <td>Q93157077</td>\n",
       "      <td>P50</td>\n",
       "      <td>Q93157070</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>566002</th>\n",
       "      <td>Q93157501-P50-1</td>\n",
       "      <td>Q93157501</td>\n",
       "      <td>P50</td>\n",
       "      <td>Q92401632</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>566003 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       id      node1  label      node2\n",
       "0       Q21090495-P2860-1  Q21090495  P2860  Q24611162\n",
       "1       Q21090495-P2860-2  Q21090495  P2860  Q24655519\n",
       "2       Q21090495-P2860-3  Q21090495  P2860  Q22065976\n",
       "3       Q21090495-P2860-4  Q21090495  P2860  Q24650035\n",
       "4       Q21090495-P2860-5  Q21090495  P2860  Q24684593\n",
       "...                   ...        ...    ...        ...\n",
       "565998    Q93147847-P50-1  Q93147847    P50  Q61104970\n",
       "565999    Q93147847-P50-2  Q93147847    P50  Q90414144\n",
       "566000    Q93147847-P50-3  Q93147847    P50  Q87706998\n",
       "566001    Q93157077-P50-1  Q93157077    P50  Q93157070\n",
       "566002    Q93157501-P50-1  Q93157501    P50  Q92401632\n",
       "\n",
       "[566003 rows x 4 columns]"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "caae = pd.read_csv(os.getenv(\"COVID\")+'/citation-and-author-edges.tsv', delimiter='\\t')\n",
    "caae.loc[:, ['id', 'node1', 'label', 'node2']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's look at some of the items we got"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[90mid\u001b[39m Q24611162\n",
      "\u001b[42mLabel\u001b[49m Viral mutation rates\n",
      "\u001b[44mDescription\u001b[49m scientific article\n",
      "\u001b[30m\u001b[47minstance of\u001b[49m\u001b[39m \u001b[90m(P31)\u001b[39m\u001b[90m: \u001b[39mscholarly article \u001b[90m(Q13442814)\u001b[39m\n",
      "\n",
      "\u001b[90mid\u001b[39m Q61104970\n",
      "\u001b[42mLabel\u001b[49m Veerasak Punyapornwithaya\n",
      "\u001b[44mDescription\u001b[49m researcher ORCID ID = 0000-0001-9870-7773\n",
      "\u001b[30m\u001b[47minstance of\u001b[49m\u001b[39m \u001b[90m(P31)\u001b[39m\u001b[90m: \u001b[39mhuman \u001b[90m(Q5)\u001b[39m\n"
     ]
    }
   ],
   "source": [
    "!wd u Q24611162, Q61104970"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Fetch all the edges from Wikidata about the authors and citations in our corpus.\n",
    "We do this by scanning the Wikidata edges file and extraction all edges where node1 matches node2 in `citation-and-author-edges.tsv`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "real\t43m13.185s\n",
      "user\t44m37.153s\n",
      "sys\t2m6.936s\n"
     ]
    }
   ],
   "source": [
    "!time gzcat \"$WD/wikidata_edges_20200504.tsv.gz\" \\\n",
    "  | kgtk ifexists --input-keys node1 --filter-on $COVID/citation-and-author-edges.tsv --filter-keys node2 \\\n",
    "  > $COVID/corpus-citations-and-authors.tsv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Keep the edges for the articles and entities that we have in the BLENDER corpus"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First extract all the edges that we may want to use (this takes 107 minutes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " 11389393 78495849 854598718 /Users/pedroszekely/data/covid/blender/corpus-citations-and-authors.tsv\n"
     ]
    }
   ],
   "source": [
    "!wc $COVID/corpus-citations-and-authors.tsv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "id\tnode1\tlabel\tnode2\trank\tnode2;magnitude\tnode2;unit\tnode2;date\tnode2;item\tnode2;lower\tnode2;upper\tnode2;latitude\tnode2;longitude\tnode2;precision\tnode2;calendar\tnode2;entity-type\n",
      "Q493567-P21-1\tQ493567\tP21\tQ6581097\tnormal\t\t\t\tQ6581097\t\t\t\t\t\t\titem\n",
      "Q493567-P106-1\tQ493567\tP106\tQ3779582\tnormal\t\t\t\tQ3779582\t\t\t\t\t\t\titem\n",
      "Q493567-P106-2\tQ493567\tP106\tQ39631\tnormal\t\t\t\tQ39631\t\t\t\t\t\t\titem\n",
      "Q493567-P106-3\tQ493567\tP106\tQ82955\tnormal\t\t\t\tQ82955\t\t\t\t\t\t\titem\n",
      "Q493567-P106-4\tQ493567\tP106\tQ1622272\tnormal\t\t\t\tQ1622272\t\t\t\t\t\t\titem\n",
      "Q493567-P106-5\tQ493567\tP106\tQ15634281\tnormal\t\t\t\tQ15634281\t\t\t\t\t\t\titem\n",
      "Q493567-P19-1\tQ493567\tP19\tQ984894\tnormal\t\t\t\tQ984894\t\t\t\t\t\t\titem\n",
      "Q493567-P244-1\tQ493567\tP244\t\"n88290830\"\tnormal\t\t\t\t\t\t\t\t\t\t\t\n",
      "Q493567-P214-1\tQ493567\tP214\t\"69131111\"\tnormal\t\t\t\t\t\t\t\t\t\t\t\n"
     ]
    }
   ],
   "source": [
    "!head $COVID/corpus-citations-and-authors.tsv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "!gzcat $COVID/corpus-identifier-edges.tsv.gz | kgtk filter -p \";P698,P932;\" | gzip > $COVID/corpus-article-identifier-edges.tsv.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "!gzcat $COVID/corpus-identifier-edges.tsv.gz | kgtk filter -p \";P685,P486,P351,P5055;\" | gzip > $COVID/corpus-entity-identifier-edges.tsv.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create Missing Items"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create Mention Edges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Incorporate Analytic Outputs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Export Knowledge Graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!time gzcat $COVID/pmcid.tsv.gz | kgtk ifexists --filter-on $COVID/covid-pmcids.tsv --left-keys node2 --right-keys id | gzip > covid-pmcid-edges.tsv.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   12934   64679  750209\n"
     ]
    }
   ],
   "source": [
    "!gzcat covid-pmcid-edges.tsv.gz | wc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "id\tnode1\tlabel\tnode2\trank\tnode2;magnitude\tnode2;unit\tnode2;lower\tnode2;upper\tnode2;latitude\tnode2;longitude\tnode2;precision\tnode2;calendar\tnode2;entity-type\n",
      "Q21093209-P932-1\tQ21093209\tP932\t\"2918564\"\tnormal\t\t\t\t\t\t\t\t\t\n",
      "Q21142689-P932-1\tQ21142689\tP932\t\"2792043\"\tnormal\t\t\t\t\t\t\t\t\t\n",
      "Q21144693-P932-1\tQ21144693\tP932\t\"1564183\"\tnormal\t\t\t\t\t\t\t\t\t\n",
      "Q21146685-P932-1\tQ21146685\tP932\t\"4207625\"\tnormal\t\t\t\t\t\t\t\t\t\n",
      "Q21256658-P932-1\tQ21256658\tP932\t\"1373654\"\tnormal\t\t\t\t\t\t\t\t\t\n",
      "Q21283966-P932-1\tQ21283966\tP932\t\"1185527\"\tnormal\t\t\t\t\t\t\t\t\t\n",
      "Q21328696-P932-1\tQ21328696\tP932\t\"2842971\"\tnormal\t\t\t\t\t\t\t\t\t\n",
      "Q22680677-P932-1\tQ22680677\tP932\t\"4517126\"\tnormal\t\t\t\t\t\t\t\t\t\n",
      "Q23912860-P932-1\tQ23912860\tP932\t\"4941879\"\tnormal\t\t\t\t\t\t\t\t\t\n",
      "gzcat: error writing to output: Broken pipe\n",
      "gzcat: covid-pmcid-edges.tsv.gz: uncompress failed\n"
     ]
    }
   ],
   "source": [
    "!gzcat covid-pmcid-edges.tsv.gz | head "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[90mid\u001b[39m Q21093209\n",
      "\u001b[42mLabel\u001b[49m RETRACTED: Influenza or not influenza: analysis of a case of high fever that happened 2000 years ago in Biblical time\n",
      "\u001b[44mDescription\u001b[49m scientific article\n",
      "\u001b[30m\u001b[47minstance of\u001b[49m\u001b[39m \u001b[90m(P31)\u001b[39m\u001b[90m: \u001b[39mscholarly article \u001b[90m(Q13442814)\u001b[39m | retracted paper \u001b[90m(Q45182324)\u001b[39m\n"
     ]
    }
   ],
   "source": [
    "!wd u Q21093209"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/pedroszekely/data/covid/covid-pmcid-edges.tsv.gz\n"
     ]
    }
   ],
   "source": [
    "!ls $COVID/covid-pmcid-edges.tsv.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "env: COVID: No such file or directory\n"
     ]
    }
   ],
   "source": [
    "!cd $(env COVID)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/Users/pedroszekely/data/covid'"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "MILLER(1)\t\t\t\t\t\t\t     MILLER(1)\n",
       "\n",
       "\n",
       "\n",
       "NAME\n",
       "       miller - like awk, sed, cut, join, and sort for name-indexed data such\n",
       "       as CSV and tabular JSON.\n",
       "\n",
       "SYNOPSIS\n",
       "       Usage: mlr [I/O options] {verb} [verb-dependent options ...] {zero or\n",
       "       more file names}\n",
       "\n",
       "\n",
       "DESCRIPTION\n",
       "       Miller operates on key-value-pair data while the familiar Unix tools\n",
       "       operate on integer-indexed fields: if the natural data structure for\n",
       "       the latter is the array, then Miller's natural data structure is the\n",
       "       insertion-ordered hash map.  This encompasses a variety of data\n",
       "       formats, including but not limited to the familiar CSV, TSV, and JSON.\n",
       "       (Miller can handle positionally-indexed data as a special case.) This\n",
       "       manpage documents Miller v5.7.0.\n",
       "\n",
       "EXAMPLES\n",
       "   COMMAND-LINE SYNTAX\n",
       "       mlr --csv cut -f hostname,uptime mydata.csv\n",
       "       mlr --tsv --rs lf filter '$status != \"down\" && $upsec >= 10000' *.tsv\n",
       "       mlr --nidx put '$sum = $7 < 0.0 ? 3.5 : $7 + 2.1*$8' *.dat\n",
       "       grep -v '^#' /etc/group | mlr --ifs : --nidx --opprint label group,pass,gid,member then sort -f group\n",
       "       mlr join -j account_id -f accounts.dat then group-by account_name balances.dat\n",
       "       mlr --json put '$attr = sub($attr, \"([0-9]+)_([0-9]+)_.*\", \"\\1:\\2\")' data/*.json\n",
       "       mlr stats1 -a min,mean,max,p10,p50,p90 -f flag,u,v data/*\n",
       "       mlr stats2 -a linreg-pca -f u,v -g shape data/*\n",
       "       mlr put -q '@sum[$a][$b] += $x; end {emit @sum, \"a\", \"b\"}' data/*\n",
       "       mlr --from estimates.tbl put '\n",
       "\t for (k,v in $*) {\n",
       "\t   if (is_numeric(v) && k =~ \"^[t-z].*$\") {\n",
       "\t     $sum += v; $count += 1\n",
       "\t   }\n",
       "\t }\n",
       "\t $mean = $sum / $count # no assignment if count unset'\n",
       "       mlr --from infile.dat put -f analyze.mlr\n",
       "       mlr --from infile.dat put 'tee > \"./taps/data-\".$a.\"-\".$b, $*'\n",
       "       mlr --from infile.dat put 'tee | \"gzip > ./taps/data-\".$a.\"-\".$b.\".gz\", $*'\n",
       "       mlr --from infile.dat put -q '@v=$*; dump | \"jq .[]\"'\n",
       "       mlr --from infile.dat put  '(NR % 1000 == 0) { print > stderr, \"Checkpoint \".NR}'\n",
       "\n",
       "   DATA FORMATS\n",
       "\t DKVP: delimited key-value pairs (Miller default format)\n",
       "\t +---------------------+\n",
       "\t | apple=1,bat=2,cog=3 | Record 1: \"apple\" => \"1\", \"bat\" => \"2\", \"cog\" => \"3\"\n",
       "\t | dish=7,egg=8,flint  | Record 2: \"dish\" => \"7\", \"egg\" => \"8\", \"3\" => \"flint\"\n",
       "\t +---------------------+\n",
       "\n",
       "\t NIDX: implicitly numerically indexed (Unix-toolkit style)\n",
       "\t +---------------------+\n",
       "\t | the quick brown     | Record 1: \"1\" => \"the\", \"2\" => \"quick\", \"3\" => \"brown\"\n",
       "\t | fox jumped\t       | Record 2: \"1\" => \"fox\", \"2\" => \"jumped\"\n",
       "\t +---------------------+\n",
       "\n",
       "\t CSV/CSV-lite: comma-separated values with separate header line\n",
       "\t +---------------------+\n",
       "\t | apple,bat,cog       |\n",
       "\t | 1,2,3\t       | Record 1: \"apple => \"1\", \"bat\" => \"2\", \"cog\" => \"3\"\n",
       "\t | 4,5,6\t       | Record 2: \"apple\" => \"4\", \"bat\" => \"5\", \"cog\" => \"6\"\n",
       "\t +---------------------+\n",
       "\n",
       "\t Tabular JSON: nested objects are supported, although arrays within them are not:\n",
       "\t +---------------------+\n",
       "\t | {\t\t       |\n",
       "\t |  \"apple\": 1,        | Record 1: \"apple\" => \"1\", \"bat\" => \"2\", \"cog\" => \"3\"\n",
       "\t |  \"bat\": 2,\t       |\n",
       "\t |  \"cog\": 3\t       |\n",
       "\t | }\t\t       |\n",
       "\t | {\t\t       |\n",
       "\t |   \"dish\": {\t       | Record 2: \"dish:egg\" => \"7\", \"dish:flint\" => \"8\", \"garlic\" => \"\"\n",
       "\t |     \"egg\": 7,       |\n",
       "\t |     \"flint\": 8      |\n",
       "\t |   }, \t       |\n",
       "\t |   \"garlic\": \"\"      |\n",
       "\t | }\t\t       |\n",
       "\t +---------------------+\n",
       "\n",
       "\t PPRINT: pretty-printed tabular\n",
       "\t +---------------------+\n",
       "\t | apple bat cog       |\n",
       "\t | 1\t 2   3\t       | Record 1: \"apple => \"1\", \"bat\" => \"2\", \"cog\" => \"3\"\n",
       "\t | 4\t 5   6\t       | Record 2: \"apple\" => \"4\", \"bat\" => \"5\", \"cog\" => \"6\"\n",
       "\t +---------------------+\n",
       "\n",
       "\t XTAB: pretty-printed transposed tabular\n",
       "\t +---------------------+\n",
       "\t | apple 1\t       | Record 1: \"apple\" => \"1\", \"bat\" => \"2\", \"cog\" => \"3\"\n",
       "\t | bat\t 2\t       |\n",
       "\t | cog\t 3\t       |\n",
       "\t |\t\t       |\n",
       "\t | dish 7\t       | Record 2: \"dish\" => \"7\", \"egg\" => \"8\"\n",
       "\t | egg\t8\t       |\n",
       "\t +---------------------+\n",
       "\n",
       "\t Markdown tabular (supported for output only):\n",
       "\t +-----------------------+\n",
       "\t | | apple | bat | cog | |\n",
       "\t | | ---   | --- | --- | |\n",
       "\t | | 1\t   | 2\t | 3   | | Record 1: \"apple => \"1\", \"bat\" => \"2\", \"cog\" => \"3\"\n",
       "\t | | 4\t   | 5\t | 6   | | Record 2: \"apple\" => \"4\", \"bat\" => \"5\", \"cog\" => \"6\"\n",
       "\t +-----------------------+\n",
       "\n",
       "OPTIONS\n",
       "       In the following option flags, the version with \"i\" designates the\n",
       "       input stream, \"o\" the output stream, and the version without prefix\n",
       "       sets the option for both input and output stream. For example: --irs\n",
       "       sets the input record separator, --ors the output record separator, and\n",
       "       --rs sets both the input and output separator to the given value.\n",
       "\n",
       "   HELP OPTIONS\n",
       "\t -h or --help\t\t      Show this message.\n",
       "\t --version\t\t      Show the software version.\n",
       "\t {verb name} --help\t      Show verb-specific help.\n",
       "\t --help-all-verbs\t      Show help on all verbs.\n",
       "\t -l or --list-all-verbs       List only verb names.\n",
       "\t -L\t\t\t      List only verb names, one per line.\n",
       "\t -f or --help-all-functions   Show help on all built-in functions.\n",
       "\t -F\t\t\t      Show a bare listing of built-in functions by name.\n",
       "\t -k or --help-all-keywords    Show help on all keywords.\n",
       "\t -K\t\t\t      Show a bare listing of keywords by name.\n",
       "\n",
       "   VERB LIST\n",
       "\taltkv bar bootstrap cat check clean-whitespace count-distinct count-similar\n",
       "\tcut decimate fill-down filter format-values fraction grep group-by group-like\n",
       "\thaving-fields head histogram join label least-frequent merge-fields\n",
       "\tmost-frequent nest nothing put regularize remove-empty-columns rename reorder\n",
       "\trepeat reshape sample sec2gmt sec2gmtdate seqgen shuffle skip-trivial-records\n",
       "\tsort stats1 stats2 step tac tail tee top uniq unsparsify\n",
       "\n",
       "   FUNCTION LIST\n",
       "\t+ + - - * / // .+ .+ .- .- .* ./ .// % ** | ^ & ~ << >> bitcount == != =~ !=~\n",
       "\t> >= < <= && || ^^ ! ? : . gsub regextract regextract_or_else strlen sub ssub\n",
       "\tsubstr tolower toupper capitalize lstrip rstrip strip collapse_whitespace\n",
       "\tclean_whitespace system abs acos acosh asin asinh atan atan2 atanh cbrt ceil\n",
       "\tcos cosh erf erfc exp expm1 floor invqnorm log log10 log1p logifit madd max\n",
       "\tmexp min mmul msub pow qnorm round roundm sgn sin sinh sqrt tan tanh urand\n",
       "\turandrange urand32 urandint dhms2fsec dhms2sec fsec2dhms fsec2hms gmt2sec\n",
       "\tlocaltime2sec hms2fsec hms2sec sec2dhms sec2gmt sec2gmt sec2gmtdate\n",
       "\tsec2localtime sec2localtime sec2localdate sec2hms strftime strftime_local\n",
       "\tstrptime strptime_local systime is_absent is_bool is_boolean is_empty\n",
       "\tis_empty_map is_float is_int is_map is_nonempty_map is_not_empty is_not_map\n",
       "\tis_not_null is_null is_numeric is_present is_string asserting_absent\n",
       "\tasserting_bool asserting_boolean asserting_empty asserting_empty_map\n",
       "\tasserting_float asserting_int asserting_map asserting_nonempty_map\n",
       "\tasserting_not_empty asserting_not_map asserting_not_null asserting_null\n",
       "\tasserting_numeric asserting_present asserting_string boolean float fmtnum\n",
       "\thexfmt int string typeof depth haskey joink joinkv joinv leafcount length\n",
       "\tmapdiff mapexcept mapselect mapsum splitkv splitkvx splitnv splitnvx\n",
       "\n",
       "       Please use \"mlr --help-function {function name}\" for function-specific help.\n",
       "\n",
       "   I/O FORMATTING\n",
       "\t --idkvp   --odkvp   --dkvp\t Delimited key-value pairs, e.g \"a=1,b=2\"\n",
       "\t\t\t\t\t (this is Miller's default format).\n",
       "\n",
       "\t --inidx   --onidx   --nidx\t Implicitly-integer-indexed fields\n",
       "\t\t\t\t\t (Unix-toolkit style).\n",
       "\t -T\t\t\t\t Synonymous with \"--nidx --fs tab\".\n",
       "\n",
       "\t --icsv    --ocsv    --csv\t Comma-separated value (or tab-separated\n",
       "\t\t\t\t\t with --fs tab, etc.)\n",
       "\n",
       "\t --itsv    --otsv    --tsv\t Keystroke-savers for \"--icsv --ifs tab\",\n",
       "\t\t\t\t\t \"--ocsv --ofs tab\", \"--csv --fs tab\".\n",
       "\t --iasv    --oasv    --asv\t Similar but using ASCII FS 0x1f and RS 0x1e\n",
       "\t --iusv    --ousv    --usv\t Similar but using Unicode FS U+241F (UTF-8 0xe2909f)\n",
       "\t\t\t\t\t and RS U+241E (UTF-8 0xe2909e)\n",
       "\n",
       "\t --icsvlite --ocsvlite --csvlite Comma-separated value (or tab-separated\n",
       "\t\t\t\t\t with --fs tab, etc.). The 'lite' CSV does not handle\n",
       "\t\t\t\t\t RFC-CSV double-quoting rules; is slightly faster;\n",
       "\t\t\t\t\t and handles heterogeneity in the input stream via\n",
       "\t\t\t\t\t empty newline followed by new header line. See also\n",
       "\t\t\t\t\t http://johnkerl.org/miller/doc/file-formats.html#CSV/TSV/etc.\n",
       "\n",
       "\t --itsvlite --otsvlite --tsvlite Keystroke-savers for \"--icsvlite --ifs tab\",\n",
       "\t\t\t\t\t \"--ocsvlite --ofs tab\", \"--csvlite --fs tab\".\n",
       "\t -t\t\t\t\t Synonymous with --tsvlite.\n",
       "\t --iasvlite --oasvlite --asvlite Similar to --itsvlite et al. but using ASCII FS 0x1f and RS 0x1e\n",
       "\t --iusvlite --ousvlite --usvlite Similar to --itsvlite et al. but using Unicode FS U+241F (UTF-8 0xe2909f)\n",
       "\t\t\t\t\t and RS U+241E (UTF-8 0xe2909e)\n",
       "\n",
       "\t --ipprint --opprint --pprint\t Pretty-printed tabular (produces no\n",
       "\t\t\t\t\t output until all input is in).\n",
       "\t\t\t     --right\t Right-justifies all fields for PPRINT output.\n",
       "\t\t\t     --barred\t Prints a border around PPRINT output\n",
       "\t\t\t\t\t (only available for output).\n",
       "\n",
       "\t\t   --omd\t\t Markdown-tabular (only available for output).\n",
       "\n",
       "\t --ixtab   --oxtab   --xtab\t Pretty-printed vertical-tabular.\n",
       "\t\t\t     --xvright\t Right-justifies values for XTAB format.\n",
       "\n",
       "\t --ijson   --ojson   --json\t JSON tabular: sequence or list of one-level\n",
       "\t\t\t\t\t maps: {...}{...} or [{...},{...}].\n",
       "\t   --json-map-arrays-on-input\t JSON arrays are unmillerable. --json-map-arrays-on-input\n",
       "\t   --json-skip-arrays-on-input\t is the default: arrays are converted to integer-indexed\n",
       "\t   --json-fatal-arrays-on-input  maps. The other two options cause them to be skipped, or\n",
       "\t\t\t\t\t to be treated as errors.  Please use the jq tool for full\n",
       "\t\t\t\t\t JSON (pre)processing.\n",
       "\t\t\t     --jvstack\t Put one key-value pair per line for JSON\n",
       "\t\t\t\t\t output.\n",
       "\t\t\t     --jlistwrap Wrap JSON output in outermost [ ].\n",
       "\t\t\t   --jknquoteint Do not quote non-string map keys in JSON output.\n",
       "\t\t\t    --jvquoteall Quote map values in JSON output, even if they're\n",
       "\t\t\t\t\t numeric.\n",
       "\t\t     --jflatsep {string} Separator for flattening multi-level JSON keys,\n",
       "\t\t\t\t\t e.g. '{\"a\":{\"b\":3}}' becomes a:b => 3 for\n",
       "\t\t\t\t\t non-JSON formats. Defaults to :.\n",
       "\n",
       "\t -p is a keystroke-saver for --nidx --fs space --repifs\n",
       "\n",
       "\t Examples: --csv for CSV-formatted input and output; --idkvp --opprint for\n",
       "\t DKVP-formatted input and pretty-printed output.\n",
       "\n",
       "\t Please use --iformat1 --oformat2 rather than --format1 --oformat2.\n",
       "\t The latter sets up input and output flags for format1, not all of which\n",
       "\t are overridden in all cases by setting output format to format2.\n",
       "\n",
       "   COMMENTS IN DATA\n",
       "\t --skip-comments\t\t Ignore commented lines (prefixed by \"#\")\n",
       "\t\t\t\t\t within the input.\n",
       "\t --skip-comments-with {string}\t Ignore commented lines within input, with\n",
       "\t\t\t\t\t specified prefix.\n",
       "\t --pass-comments\t\t Immediately print commented lines (prefixed by \"#\")\n",
       "\t\t\t\t\t within the input.\n",
       "\t --pass-comments-with {string}\t Immediately print commented lines within input, with\n",
       "\t\t\t\t\t specified prefix.\n",
       "       Notes:\n",
       "       * Comments are only honored at the start of a line.\n",
       "       * In the absence of any of the above four options, comments are data like\n",
       "\t any other text.\n",
       "       * When pass-comments is used, comment lines are written to standard output\n",
       "\t immediately upon being read; they are not part of the record stream.\n",
       "\t Results may be counterintuitive. A suggestion is to place comments at the\n",
       "\t start of data files.\n",
       "\n",
       "   FORMAT-CONVERSION KEYSTROKE-SAVERS\n",
       "       As keystroke-savers for format-conversion you may use the following:\n",
       "\t       --c2t --c2d --c2n --c2j --c2x --c2p --c2m\n",
       "\t --t2c\t     --t2d --t2n --t2j --t2x --t2p --t2m\n",
       "\t --d2c --d2t\t   --d2n --d2j --d2x --d2p --d2m\n",
       "\t --n2c --n2t --n2d\t --n2j --n2x --n2p --n2m\n",
       "\t --j2c --j2t --j2d --j2n       --j2x --j2p --j2m\n",
       "\t --x2c --x2t --x2d --x2n --x2j\t     --x2p --x2m\n",
       "\t --p2c --p2t --p2d --p2n --p2j --p2x\t   --p2m\n",
       "       The letters c t d n j x p m refer to formats CSV, TSV, DKVP, NIDX, JSON, XTAB,\n",
       "       PPRINT, and markdown, respectively. Note that markdown format is available for\n",
       "       output only.\n",
       "\n",
       "   COMPRESSED I/O\n",
       "\t --prepipe {command} This allows Miller to handle compressed inputs. You can do\n",
       "\t without this for single input files, e.g. \"gunzip < myfile.csv.gz | mlr ...\".\n",
       "\t However, when multiple input files are present, between-file separations are\n",
       "\t lost; also, the FILENAME variable doesn't iterate. Using --prepipe you can\n",
       "\t specify an action to be taken on each input file. This pre-pipe command must\n",
       "\t be able to read from standard input; it will be invoked with\n",
       "\t   {command} < {filename}.\n",
       "\t Examples:\n",
       "\t   mlr --prepipe 'gunzip'\n",
       "\t   mlr --prepipe 'zcat -cf'\n",
       "\t   mlr --prepipe 'xz -cd'\n",
       "\t   mlr --prepipe cat\n",
       "\t Note that this feature is quite general and is not limited to decompression\n",
       "\t utilities. You can use it to apply per-file filters of your choice.\n",
       "\t For output compression (or other) utilities, simply pipe the output:\n",
       "\t   mlr ... | {your compression command}\n",
       "\n",
       "   SEPARATORS\n",
       "\t --rs\t  --irs     --ors\t       Record separators, e.g. 'lf' or '\\r\\n'\n",
       "\t --fs\t  --ifs     --ofs  --repifs    Field separators, e.g. comma\n",
       "\t --ps\t  --ips     --ops\t       Pair separators, e.g. equals sign\n",
       "\n",
       "\t Notes about line endings:\n",
       "\t * Default line endings (--irs and --ors) are \"auto\" which means autodetect from\n",
       "\t   the input file format, as long as the input file(s) have lines ending in either\n",
       "\t   LF (also known as linefeed, '\\n', 0x0a, Unix-style) or CRLF (also known as\n",
       "\t   carriage-return/linefeed pairs, '\\r\\n', 0x0d 0x0a, Windows style).\n",
       "\t * If both irs and ors are auto (which is the default) then LF input will lead to LF\n",
       "\t   output and CRLF input will lead to CRLF output, regardless of the platform you're\n",
       "\t   running on.\n",
       "\t * The line-ending autodetector triggers on the first line ending detected in the input\n",
       "\t   stream. E.g. if you specify a CRLF-terminated file on the command line followed by an\n",
       "\t   LF-terminated file then autodetected line endings will be CRLF.\n",
       "\t * If you use --ors {something else} with (default or explicitly specified) --irs auto\n",
       "\t   then line endings are autodetected on input and set to what you specify on output.\n",
       "\t * If you use --irs {something else} with (default or explicitly specified) --ors auto\n",
       "\t   then the output line endings used are LF on Unix/Linux/BSD/MacOSX, and CRLF on Windows.\n",
       "\n",
       "\t Notes about all other separators:\n",
       "\t * IPS/OPS are only used for DKVP and XTAB formats, since only in these formats\n",
       "\t   do key-value pairs appear juxtaposed.\n",
       "\t * IRS/ORS are ignored for XTAB format. Nominally IFS and OFS are newlines;\n",
       "\t   XTAB records are separated by two or more consecutive IFS/OFS -- i.e.\n",
       "\t   a blank line. Everything above about --irs/--ors/--rs auto becomes --ifs/--ofs/--fs\n",
       "\t   auto for XTAB format. (XTAB's default IFS/OFS are \"auto\".)\n",
       "\t * OFS must be single-character for PPRINT format. This is because it is used\n",
       "\t   with repetition for alignment; multi-character separators would make\n",
       "\t   alignment impossible.\n",
       "\t * OPS may be multi-character for XTAB format, in which case alignment is\n",
       "\t   disabled.\n",
       "\t * TSV is simply CSV using tab as field separator (\"--fs tab\").\n",
       "\t * FS/PS are ignored for markdown format; RS is used.\n",
       "\t * All FS and PS options are ignored for JSON format, since they are not relevant\n",
       "\t   to the JSON format.\n",
       "\t * You can specify separators in any of the following ways, shown by example:\n",
       "\t   - Type them out, quoting as necessary for shell escapes, e.g.\n",
       "\t     \"--fs '|' --ips :\"\n",
       "\t   - C-style escape sequences, e.g. \"--rs '\\r\\n' --fs '\\t'\".\n",
       "\t   - To avoid backslashing, you can use any of the following names:\n",
       "\t     cr crcr newline lf lflf crlf crlfcrlf tab space comma pipe slash colon semicolon equals\n",
       "\t * Default separators by format:\n",
       "\t     File format  RS\t   FS\t    PS\n",
       "\t     gen\t  N/A\t   (N/A)    (N/A)\n",
       "\t     dkvp\t  auto\t   ,\t    =\n",
       "\t     json\t  auto\t   (N/A)    (N/A)\n",
       "\t     nidx\t  auto\t   space    (N/A)\n",
       "\t     csv\t  auto\t   ,\t    (N/A)\n",
       "\t     csvlite\t  auto\t   ,\t    (N/A)\n",
       "\t     markdown\t  auto\t   (N/A)    (N/A)\n",
       "\t     pprint\t  auto\t   space    (N/A)\n",
       "\t     xtab\t  (N/A)    auto     space\n",
       "\n",
       "   CSV-SPECIFIC OPTIONS\n",
       "\t --implicit-csv-header Use 1,2,3,... as field labels, rather than from line 1\n",
       "\t\t\t    of input files. Tip: combine with \"label\" to recreate\n",
       "\t\t\t    missing headers.\n",
       "\t --allow-ragged-csv-input|--ragged If a data line has fewer fields than the header line,\n",
       "\t\t\t    fill remaining keys with empty string. If a data line has more\n",
       "\t\t\t    fields than the header line, use integer field labels as in\n",
       "\t\t\t    the implicit-header case.\n",
       "\t --headerless-csv-output   Print only CSV data lines.\n",
       "\t -N\t\t    Keystroke-saver for --implicit-csv-header --headerless-csv-output.\n",
       "\n",
       "   DOUBLE-QUOTING FOR CSV/CSVLITE OUTPUT\n",
       "\t --quote-all\t    Wrap all fields in double quotes\n",
       "\t --quote-none\t    Do not wrap any fields in double quotes, even if they have\n",
       "\t\t\t    OFS or ORS in them\n",
       "\t --quote-minimal    Wrap fields in double quotes only if they have OFS or ORS\n",
       "\t\t\t    in them (default)\n",
       "\t --quote-numeric    Wrap fields in double quotes only if they have numbers\n",
       "\t\t\t    in them\n",
       "\t --quote-original   Wrap fields in double quotes if and only if they were\n",
       "\t\t\t    quoted on input. This isn't sticky for computed fields:\n",
       "\t\t\t    e.g. if fields a and b were quoted on input and you do\n",
       "\t\t\t    \"put '$c = $a . $b'\" then field c won't inherit a or b's\n",
       "\t\t\t    was-quoted-on-input flag.\n",
       "\n",
       "   NUMERICAL FORMATTING\n",
       "\t --ofmt {format}    E.g. %.18lf, %.0lf. Please use sprintf-style codes for\n",
       "\t\t\t    double-precision. Applies to verbs which compute new\n",
       "\t\t\t    values, e.g. put, stats1, stats2. See also the fmtnum\n",
       "\t\t\t    function within mlr put (mlr --help-all-functions).\n",
       "\t\t\t    Defaults to %lf.\n",
       "\n",
       "   OTHER OPTIONS\n",
       "\t --seed {n} with n of the form 12345678 or 0xcafefeed. For put/filter\n",
       "\t\t\t    urand()/urandint()/urand32().\n",
       "\t --nr-progress-mod {m}, with m a positive integer: print filename and record\n",
       "\t\t\t    count to stderr every m input records.\n",
       "\t --from {filename}  Use this to specify an input file before the verb(s),\n",
       "\t\t\t    rather than after. May be used more than once. Example:\n",
       "\t\t\t    \"mlr --from a.dat --from b.dat cat\" is the same as\n",
       "\t\t\t    \"mlr cat a.dat b.dat\".\n",
       "\t -n\t\t    Process no input files, nor standard input either. Useful\n",
       "\t\t\t    for mlr put with begin/end statements only. (Same as --from\n",
       "\t\t\t    /dev/null.) Also useful in \"mlr -n put -v '...'\" for\n",
       "\t\t\t    analyzing abstract syntax trees (if that's your thing).\n",
       "\t -I\t\t    Process files in-place. For each file name on the command\n",
       "\t\t\t    line, output is written to a temp file in the same\n",
       "\t\t\t    directory, which is then renamed over the original. Each\n",
       "\t\t\t    file is processed in isolation: if the output format is\n",
       "\t\t\t    CSV, CSV headers will be present in each output file;\n",
       "\t\t\t    statistics are only over each file's own records; and so on.\n",
       "\n",
       "   THEN-CHAINING\n",
       "       Output of one verb may be chained as input to another using \"then\", e.g.\n",
       "\t mlr stats1 -a min,mean,max -f flag,u,v -g color then sort -f color\n",
       "\n",
       "   AUXILIARY COMMANDS\n",
       "       Miller has a few otherwise-standalone executables packaged within it.\n",
       "       They do not participate in any other parts of Miller.\n",
       "       Available subcommands:\n",
       "\t aux-list\n",
       "\t lecat\n",
       "\t termcvt\n",
       "\t hex\n",
       "\t unhex\n",
       "\t netbsd-strptime\n",
       "       For more information, please invoke mlr {subcommand} --help\n",
       "\n",
       "VERBS\n",
       "   altkv\n",
       "       Usage: mlr altkv [no options]\n",
       "       Given fields with values of the form a,b,c,d,e,f emits a=b,c=d,e=f pairs.\n",
       "\n",
       "   bar\n",
       "       Usage: mlr bar [options]\n",
       "       Replaces a numeric field with a number of asterisks, allowing for cheesy\n",
       "       bar plots. These align best with --opprint or --oxtab output format.\n",
       "       Options:\n",
       "       -f   {a,b,c}\t Field names to convert to bars.\n",
       "       -c   {character}  Fill character: default '*'.\n",
       "       -x   {character}  Out-of-bounds character: default '#'.\n",
       "       -b   {character}  Blank character: default '.'.\n",
       "       --lo {lo}\t Lower-limit value for min-width bar: default '0.000000'.\n",
       "       --hi {hi}\t Upper-limit value for max-width bar: default '100.000000'.\n",
       "       -w   {n} \t Bar-field width: default '40'.\n",
       "       --auto\t\t Automatically computes limits, ignoring --lo and --hi.\n",
       "\t\t\t Holds all records in memory before producing any output.\n",
       "\n",
       "   bootstrap\n",
       "       Usage: mlr bootstrap [options]\n",
       "       Emits an n-sample, with replacement, of the input records.\n",
       "       Options:\n",
       "       -n {number} Number of samples to output. Defaults to number of input records.\n",
       "\t\t   Must be non-negative.\n",
       "       See also mlr sample and mlr shuffle.\n",
       "\n",
       "   cat\n",
       "       Usage: mlr cat [options]\n",
       "       Passes input records directly to output. Most useful for format conversion.\n",
       "       Options:\n",
       "       -n\t Prepend field \"n\" to each record with record-counter starting at 1\n",
       "       -g {comma-separated field name(s)} When used with -n/-N, writes record-counters\n",
       "\t\t keyed by specified field name(s).\n",
       "       -v\t Write a low-level record-structure dump to stderr.\n",
       "       -N {name} Prepend field {name} to each record with record-counter starting at 1\n",
       "\n",
       "   check\n",
       "       Usage: mlr check\n",
       "       Consumes records without printing any output.\n",
       "       Useful for doing a well-formatted check on input data.\n",
       "\n",
       "   clean-whitespace\n",
       "       Usage: mlr clean-whitespace [options] {old1,new1,old2,new2,...}\n",
       "       For each record, for each field in the record, whitespace-cleans the keys and\n",
       "       values. Whitespace-cleaning entails stripping leading and trailing whitespace,\n",
       "       and replacing multiple whitespace with singles. For finer-grained control,\n",
       "       please see the DSL functions lstrip, rstrip, strip, collapse_whitespace,\n",
       "       and clean_whitespace.\n",
       "\n",
       "       Options:\n",
       "       -k|--keys-only\t Do not touch values.\n",
       "       -v|--values-only  Do not touch keys.\n",
       "       It is an error to specify -k as well as -v.\n",
       "\n",
       "   count-distinct\n",
       "       Usage: mlr count-distinct [options]\n",
       "       Prints number of records having distinct values for specified field names.\n",
       "       Same as uniq -c.\n",
       "\n",
       "       Options:\n",
       "       -f {a,b,c}    Field names for distinct count.\n",
       "       -n\t     Show only the number of distinct values. Not compatible with -u.\n",
       "       -o {name}     Field name for output count. Default \"count\".\n",
       "\t\t     Ignored with -u.\n",
       "       -u\t     Do unlashed counts for multiple field names. With -f a,b and\n",
       "\t\t     without -u, computes counts for distinct combinations of a\n",
       "\t\t     and b field values. With -f a,b and with -u, computes counts\n",
       "\t\t     for distinct a field values and counts for distinct b field\n",
       "\t\t     values separately.\n",
       "\n",
       "   count-similar\n",
       "       Usage: mlr count-similar [options]\n",
       "       Ingests all records, then emits each record augmented by a count of\n",
       "       the number of other records having the same group-by field values.\n",
       "       Options:\n",
       "       -g {d,e,f} Group-by-field names for counts.\n",
       "       -o {name}  Field name for output count. Default \"count\".\n",
       "\n",
       "   cut\n",
       "       Usage: mlr cut [options]\n",
       "       Passes through input records with specified fields included/excluded.\n",
       "       -f {a,b,c}\tField names to include for cut.\n",
       "       -o\t\tRetain fields in the order specified here in the argument list.\n",
       "\t\t\tDefault is to retain them in the order found in the input data.\n",
       "       -x|--complement\tExclude, rather than include, field names specified by -f.\n",
       "       -r\t\tTreat field names as regular expressions. \"ab\", \"a.*b\" will\n",
       "\t\t\tmatch any field name containing the substring \"ab\" or matching\n",
       "\t\t\t\"a.*b\", respectively; anchors of the form \"^ab$\", \"^a.*b$\" may\n",
       "\t\t\tbe used. The -o flag is ignored when -r is present.\n",
       "       Examples:\n",
       "\t mlr cut -f hostname,status\n",
       "\t mlr cut -x -f hostname,status\n",
       "\t mlr cut -r -f '^status$,sda[0-9]'\n",
       "\t mlr cut -r -f '^status$,\"sda[0-9]\"'\n",
       "\t mlr cut -r -f '^status$,\"sda[0-9]\"i' (this is case-insensitive)\n",
       "\n",
       "   decimate\n",
       "       Usage: mlr decimate [options]\n",
       "       -n {count}    Decimation factor; default 10\n",
       "       -b\t     Decimate by printing first of every n.\n",
       "       -e\t     Decimate by printing last of every n (default).\n",
       "       -g {a,b,c}    Optional group-by-field names for decimate counts\n",
       "       Passes through one of every n records, optionally by category.\n",
       "\n",
       "   fill-down\n",
       "       Usage: mlr fill-down [options]\n",
       "       -f {a,b,c}\t   Field names for fill-down\n",
       "       -a|--only-if-absent Field names for fill-down\n",
       "       If a given record has a missing value for a given field, fill that from\n",
       "       the corresponding value from a previous record, if any.\n",
       "       By default, a 'missing' field either is absent, or has the empty-string value.\n",
       "       With -a, a field is 'missing' only if it is absent.\n",
       "\n",
       "   filter\n",
       "       Usage: mlr filter [options] {expression}\n",
       "       Prints records for which {expression} evaluates to true.\n",
       "       If there are multiple semicolon-delimited expressions, all of them are\n",
       "       evaluated and the last one is used as the filter criterion.\n",
       "\n",
       "       Conversion options:\n",
       "       -S: Keeps field values as strings with no type inference to int or float.\n",
       "       -F: Keeps field values as strings or floats with no inference to int.\n",
       "       All field values are type-inferred to int/float/string unless this behavior is\n",
       "       suppressed with -S or -F.\n",
       "\n",
       "       Output/formatting options:\n",
       "       --oflatsep {string}: Separator to use when flattening multi-level @-variables\n",
       "\t   to output records for emit. Default \":\".\n",
       "       --jknquoteint: For dump output (JSON-formatted), do not quote map keys if non-string.\n",
       "       --jvquoteall: For dump output (JSON-formatted), quote map values even if non-string.\n",
       "       Any of the output-format command-line flags (see mlr -h). Example: using\n",
       "\t mlr --icsv --opprint ... then put --ojson 'tee > \"mytap-\".$a.\".dat\", $*' then ...\n",
       "       the input is CSV, the output is pretty-print tabular, but the tee-file output\n",
       "       is written in JSON format.\n",
       "       --no-fflush: for emit, tee, print, and dump, don't call fflush() after every\n",
       "\t   record.\n",
       "\n",
       "       Expression-specification options:\n",
       "       -f {filename}: the DSL expression is taken from the specified file rather\n",
       "\t   than from the command line. Outer single quotes wrapping the expression\n",
       "\t   should not be placed in the file. If -f is specified more than once,\n",
       "\t   all input files specified using -f are concatenated to produce the expression.\n",
       "\t   (For example, you can define functions in one file and call them from another.)\n",
       "       -e {expression}: You can use this after -f to add an expression. Example use\n",
       "\t   case: define functions/subroutines in a file you specify with -f, then call\n",
       "\t   them with an expression you specify with -e.\n",
       "       (If you mix -e and -f then the expressions are evaluated in the order encountered.\n",
       "       Since the expression pieces are simply concatenated, please be sure to use intervening\n",
       "       semicolons to separate expressions.)\n",
       "\n",
       "       -s name=value: Predefines out-of-stream variable @name to have value \"value\".\n",
       "\t   Thus mlr filter put -s foo=97 '$column += @foo' is like\n",
       "\t   mlr filter put 'begin {@foo = 97} $column += @foo'.\n",
       "\t   The value part is subject to type-inferencing as specified by -S/-F.\n",
       "\t   May be specified more than once, e.g. -s name1=value1 -s name2=value2.\n",
       "\n",
       "       Tracing options:\n",
       "       -v: Prints the expressions's AST (abstract syntax tree), which gives\n",
       "\t   full transparency on the precedence and associativity rules of\n",
       "\t   Miller's grammar, to stdout.\n",
       "       -a: Prints a low-level stack-allocation trace to stdout.\n",
       "       -t: Prints a low-level parser trace to stderr.\n",
       "       -T: Prints a every statement to stderr as it is executed.\n",
       "\n",
       "       Other options:\n",
       "       -x: Prints records for which {expression} evaluates to false.\n",
       "\n",
       "       Please use a dollar sign for field names and double-quotes for string\n",
       "       literals. If field names have special characters such as \".\" then you might\n",
       "       use braces, e.g. '${field.name}'. Miller built-in variables are\n",
       "       NF NR FNR FILENUM FILENAME M_PI M_E, and ENV[\"namegoeshere\"] to access environment\n",
       "       variables. The environment-variable name may be an expression, e.g. a field\n",
       "       value.\n",
       "\n",
       "       Use # to comment to end of line.\n",
       "\n",
       "       Examples:\n",
       "\t mlr filter 'log10($count) > 4.0'\n",
       "\t mlr filter 'FNR == 2\t       (second record in each file)'\n",
       "\t mlr filter 'urand() < 0.001'  (subsampling)\n",
       "\t mlr filter '$color != \"blue\" && $value > 4.2'\n",
       "\t mlr filter '($x<.5 && $y<.5) || ($x>.5 && $y>.5)'\n",
       "\t mlr filter '($name =~ \"^sys.*east$\") || ($name =~ \"^dev.[0-9]+\"i)'\n",
       "\t mlr filter '$ab = $a+$b; $cd = $c+$d; $ab != $cd'\n",
       "\t mlr filter '\n",
       "\t   NR == 1 ||\n",
       "\t  #NR == 2 ||\n",
       "\t   NR == 3\n",
       "\t '\n",
       "\n",
       "       Please see http://johnkerl.org/miller/doc/reference.html for more information\n",
       "       including function list. Or \"mlr -f\". Please also also \"mlr grep\" which is\n",
       "       useful when you don't yet know which field name(s) you're looking for.\n",
       "\n",
       "   format-values\n",
       "       Usage: mlr format-values [options]\n",
       "       Applies format strings to all field values, depending on autodetected type.\n",
       "       * If a field value is detected to be integer, applies integer format.\n",
       "       * Else, if a field value is detected to be float, applies float format.\n",
       "       * Else, applies string format.\n",
       "\n",
       "       Note: this is a low-keystroke way to apply formatting to many fields. To get\n",
       "       finer control, please see the fmtnum function within the mlr put DSL.\n",
       "\n",
       "       Note: this verb lets you apply arbitrary format strings, which can produce\n",
       "       undefined behavior and/or program crashes.  See your system's \"man printf\".\n",
       "\n",
       "       Options:\n",
       "       -i {integer format} Defaults to \"%lld\".\n",
       "\t\t\t   Examples: \"%06lld\", \"%08llx\".\n",
       "\t\t\t   Note that Miller integers are long long so you must use\n",
       "\t\t\t   formats which apply to long long, e.g. with ll in them.\n",
       "\t\t\t   Undefined behavior results otherwise.\n",
       "       -f {float format}   Defaults to \"%lf\".\n",
       "\t\t\t   Examples: \"%8.3lf\", \"%.6le\".\n",
       "\t\t\t   Note that Miller floats are double-precision so you must\n",
       "\t\t\t   use formats which apply to double, e.g. with l[efg] in them.\n",
       "\t\t\t   Undefined behavior results otherwise.\n",
       "       -s {string format}  Defaults to \"%s\".\n",
       "\t\t\t   Examples: \"_%s\", \"%08s\".\n",
       "\t\t\t   Note that you must use formats which apply to string, e.g.\n",
       "\t\t\t   with s in them. Undefined behavior results otherwise.\n",
       "       -n\t\t   Coerce field values autodetected as int to float, and then\n",
       "\t\t\t   apply the float format.\n",
       "\n",
       "   fraction\n",
       "       Usage: mlr fraction [options]\n",
       "       For each record's value in specified fields, computes the ratio of that\n",
       "       value to the sum of values in that field over all input records.\n",
       "       E.g. with input records\tx=1  x=2  x=3  and  x=4, emits output records\n",
       "       x=1,x_fraction=0.1  x=2,x_fraction=0.2  x=3,x_fraction=0.3  and\tx=4,x_fraction=0.4\n",
       "\n",
       "       Note: this is internally a two-pass algorithm: on the first pass it retains\n",
       "       input records and accumulates sums; on the second pass it computes quotients\n",
       "       and emits output records. This means it produces no output until all input is read.\n",
       "\n",
       "       Options:\n",
       "       -f {a,b,c}    Field name(s) for fraction calculation\n",
       "       -g {d,e,f}    Optional group-by-field name(s) for fraction counts\n",
       "       -p\t     Produce percents [0..100], not fractions [0..1]. Output field names\n",
       "\t\t     end with \"_percent\" rather than \"_fraction\"\n",
       "       -c\t     Produce cumulative distributions, i.e. running sums: each output\n",
       "\t\t     value folds in the sum of the previous for the specified group\n",
       "\t\t     E.g. with input records  x=1  x=2\tx=3  and  x=4, emits output records\n",
       "\t\t     x=1,x_cumulative_fraction=0.1  x=2,x_cumulative_fraction=0.3\n",
       "\t\t     x=3,x_cumulative_fraction=0.6  and  x=4,x_cumulative_fraction=1.0\n",
       "\n",
       "   grep\n",
       "       Usage: mlr grep [options] {regular expression}\n",
       "       Passes through records which match {regex}.\n",
       "       Options:\n",
       "       -i    Use case-insensitive search.\n",
       "       -v    Invert: pass through records which do not match the regex.\n",
       "       Note that \"mlr filter\" is more powerful, but requires you to know field names.\n",
       "       By contrast, \"mlr grep\" allows you to regex-match the entire record. It does\n",
       "       this by formatting each record in memory as DKVP, using command-line-specified\n",
       "       ORS/OFS/OPS, and matching the resulting line against the regex specified\n",
       "       here. In particular, the regex is not applied to the input stream: if you\n",
       "       have CSV with header line \"x,y,z\" and data line \"1,2,3\" then the regex will\n",
       "       be matched, not against either of these lines, but against the DKVP line\n",
       "       \"x=1,y=2,z=3\".  Furthermore, not all the options to system grep are supported,\n",
       "       and this command is intended to be merely a keystroke-saver. To get all the\n",
       "       features of system grep, you can do\n",
       "\t \"mlr --odkvp ... | grep ... | mlr --idkvp ...\"\n",
       "\n",
       "   group-by\n",
       "       Usage: mlr group-by {comma-separated field names}\n",
       "       Outputs records in batches having identical values at specified field names.\n",
       "\n",
       "   group-like\n",
       "       Usage: mlr group-like\n",
       "       Outputs records in batches having identical field names.\n",
       "\n",
       "   having-fields\n",
       "       Usage: mlr having-fields [options]\n",
       "       Conditionally passes through records depending on each record's field names.\n",
       "       Options:\n",
       "\t --at-least\t {comma-separated names}\n",
       "\t --which-are\t {comma-separated names}\n",
       "\t --at-most\t {comma-separated names}\n",
       "\t --all-matching  {regular expression}\n",
       "\t --any-matching  {regular expression}\n",
       "\t --none-matching {regular expression}\n",
       "       Examples:\n",
       "\t mlr having-fields --which-are amount,status,owner\n",
       "\t mlr having-fields --any-matching 'sda[0-9]'\n",
       "\t mlr having-fields --any-matching '\"sda[0-9]\"'\n",
       "\t mlr having-fields --any-matching '\"sda[0-9]\"i' (this is case-insensitive)\n",
       "\n",
       "   head\n",
       "       Usage: mlr head [options]\n",
       "       -n {count}    Head count to print; default 10\n",
       "       -g {a,b,c}    Optional group-by-field names for head counts\n",
       "       Passes through the first n records, optionally by category.\n",
       "       Without -g, ceases consuming more input (i.e. is fast) when n\n",
       "       records have been read.\n",
       "\n",
       "   histogram\n",
       "       Usage: mlr histogram [options]\n",
       "       -f {a,b,c}    Value-field names for histogram counts\n",
       "       --lo {lo}     Histogram low value\n",
       "       --hi {hi}     Histogram high value\n",
       "       --nbins {n}   Number of histogram bins\n",
       "       --auto\t     Automatically computes limits, ignoring --lo and --hi.\n",
       "\t\t     Holds all values in memory before producing any output.\n",
       "       -o {prefix}   Prefix for output field name. Default: no prefix.\n",
       "       Just a histogram. Input values < lo or > hi are not counted.\n",
       "\n",
       "   join\n",
       "       Usage: mlr join [options]\n",
       "       Joins records from specified left file name with records from all file names\n",
       "       at the end of the Miller argument list.\n",
       "       Functionality is essentially the same as the system \"join\" command, but for\n",
       "       record streams.\n",
       "       Options:\n",
       "\t -f {left file name}\n",
       "\t -j {a,b,c}   Comma-separated join-field names for output\n",
       "\t -l {a,b,c}   Comma-separated join-field names for left input file;\n",
       "\t\t      defaults to -j values if omitted.\n",
       "\t -r {a,b,c}   Comma-separated join-field names for right input file(s);\n",
       "\t\t      defaults to -j values if omitted.\n",
       "\t --lp {text}  Additional prefix for non-join output field names from\n",
       "\t\t      the left file\n",
       "\t --rp {text}  Additional prefix for non-join output field names from\n",
       "\t\t      the right file(s)\n",
       "\t --np\t      Do not emit paired records\n",
       "\t --ul\t      Emit unpaired records from the left file\n",
       "\t --ur\t      Emit unpaired records from the right file(s)\n",
       "\t -s|--sorted-input  Require sorted input: records must be sorted\n",
       "\t\t      lexically by their join-field names, else not all records will\n",
       "\t\t      be paired. The only likely use case for this is with a left\n",
       "\t\t      file which is too big to fit into system memory otherwise.\n",
       "\t -u\t      Enable unsorted input. (This is the default even without -u.)\n",
       "\t\t      In this case, the entire left file will be loaded into memory.\n",
       "\t --prepipe {command} As in main input options; see mlr --help for details.\n",
       "\t\t      If you wish to use a prepipe command for the main input as well\n",
       "\t\t      as here, it must be specified there as well as here.\n",
       "       File-format options default to those for the right file names on the Miller\n",
       "       argument list, but may be overridden for the left file as follows. Please see\n",
       "       the main \"mlr --help\" for more information on syntax for these arguments.\n",
       "\t -i {one of csv,dkvp,nidx,pprint,xtab}\n",
       "\t --irs {record-separator character}\n",
       "\t --ifs {field-separator character}\n",
       "\t --ips {pair-separator character}\n",
       "\t --repifs\n",
       "\t --repips\n",
       "       Please use \"mlr --usage-separator-options\" for information on specifying separators.\n",
       "       Please see http://johnkerl.org/miller/doc/reference.html for more information\n",
       "       including examples.\n",
       "\n",
       "   label\n",
       "       Usage: mlr label {new1,new2,new3,...}\n",
       "       Given n comma-separated names, renames the first n fields of each record to\n",
       "       have the respective name. (Fields past the nth are left with their original\n",
       "       names.) Particularly useful with --inidx or --implicit-csv-header, to give\n",
       "       useful names to otherwise integer-indexed fields.\n",
       "       Examples:\n",
       "\t \"echo 'a b c d' | mlr --inidx --odkvp cat\"\t  gives \"1=a,2=b,3=c,4=d\"\n",
       "\t \"echo 'a b c d' | mlr --inidx --odkvp label s,t\" gives \"s=a,t=b,3=c,4=d\"\n",
       "\n",
       "   least-frequent\n",
       "       Usage: mlr least-frequent [options]\n",
       "       Shows the least frequently occurring distinct values for specified field names.\n",
       "       The first entry is the statistical anti-mode; the remaining are runners-up.\n",
       "       Options:\n",
       "       -f {one or more comma-separated field names}. Required flag.\n",
       "       -n {count}. Optional flag defaulting to 10.\n",
       "       -b\t   Suppress counts; show only field values.\n",
       "       -o {name}   Field name for output count. Default \"count\".\n",
       "       See also \"mlr most-frequent\".\n",
       "\n",
       "   merge-fields\n",
       "       Usage: mlr merge-fields [options]\n",
       "       Computes univariate statistics for each input record, accumulated across\n",
       "       specified fields.\n",
       "       Options:\n",
       "       -a {sum,count,...}  Names of accumulators. One or more of:\n",
       "\t count\t   Count instances of fields\n",
       "\t mode\t   Find most-frequently-occurring values for fields; first-found wins tie\n",
       "\t antimode  Find least-frequently-occurring values for fields; first-found wins tie\n",
       "\t sum\t   Compute sums of specified fields\n",
       "\t mean\t   Compute averages (sample means) of specified fields\n",
       "\t stddev    Compute sample standard deviation of specified fields\n",
       "\t var\t   Compute sample variance of specified fields\n",
       "\t meaneb    Estimate error bars for averages (assuming no sample autocorrelation)\n",
       "\t skewness  Compute sample skewness of specified fields\n",
       "\t kurtosis  Compute sample kurtosis of specified fields\n",
       "\t min\t   Compute minimum values of specified fields\n",
       "\t max\t   Compute maximum values of specified fields\n",
       "       -f {a,b,c}  Value-field names on which to compute statistics. Requires -o.\n",
       "       -r {a,b,c}  Regular expressions for value-field names on which to compute\n",
       "\t\t   statistics. Requires -o.\n",
       "       -c {a,b,c}  Substrings for collapse mode. All fields which have the same names\n",
       "\t\t   after removing substrings will be accumulated together. Please see\n",
       "\t\t   examples below.\n",
       "       -i\t   Use interpolated percentiles, like R's type=7; default like type=1.\n",
       "\t\t   Not sensical for string-valued fields.\n",
       "       -o {name}   Output field basename for -f/-r.\n",
       "       -k\t   Keep the input fields which contributed to the output statistics;\n",
       "\t\t   the default is to omit them.\n",
       "       -F\t   Computes integerable things (e.g. count) in floating point.\n",
       "\n",
       "       String-valued data make sense unless arithmetic on them is required,\n",
       "       e.g. for sum, mean, interpolated percentiles, etc. In case of mixed data,\n",
       "       numbers are less than strings.\n",
       "\n",
       "       Example input data: \"a_in_x=1,a_out_x=2,b_in_y=4,b_out_x=8\".\n",
       "       Example: mlr merge-fields -a sum,count -f a_in_x,a_out_x -o foo\n",
       "\t produces \"b_in_y=4,b_out_x=8,foo_sum=3,foo_count=2\" since \"a_in_x,a_out_x\" are\n",
       "\t summed over.\n",
       "       Example: mlr merge-fields -a sum,count -r in_,out_ -o bar\n",
       "\t produces \"bar_sum=15,bar_count=4\" since all four fields are summed over.\n",
       "       Example: mlr merge-fields -a sum,count -c in_,out_\n",
       "\t produces \"a_x_sum=3,a_x_count=2,b_y_sum=4,b_y_count=1,b_x_sum=8,b_x_count=1\"\n",
       "\t since \"a_in_x\" and \"a_out_x\" both collapse to \"a_x\", \"b_in_y\" collapses to\n",
       "\t \"b_y\", and \"b_out_x\" collapses to \"b_x\".\n",
       "\n",
       "   most-frequent\n",
       "       Usage: mlr most-frequent [options]\n",
       "       Shows the most frequently occurring distinct values for specified field names.\n",
       "       The first entry is the statistical mode; the remaining are runners-up.\n",
       "       Options:\n",
       "       -f {one or more comma-separated field names}. Required flag.\n",
       "       -n {count}. Optional flag defaulting to 10.\n",
       "       -b\t   Suppress counts; show only field values.\n",
       "       -o {name}   Field name for output count. Default \"count\".\n",
       "       See also \"mlr least-frequent\".\n",
       "\n",
       "   nest\n",
       "       Usage: mlr nest [options]\n",
       "       Explodes specified field values into separate fields/records, or reverses this.\n",
       "       Options:\n",
       "\t --explode,--implode   One is required.\n",
       "\t --values,--pairs      One is required.\n",
       "\t --across-records,--across-fields One is required.\n",
       "\t -f {field name}       Required.\n",
       "\t --nested-fs {string}  Defaults to \";\". Field separator for nested values.\n",
       "\t --nested-ps {string}  Defaults to \":\". Pair separator for nested key-value pairs.\n",
       "\t --evar {string}       Shorthand for --explode --values ---across-records --nested-fs {string}\n",
       "\t --ivar {string}       Shorthand for --implode --values ---across-records --nested-fs {string}\n",
       "       Please use \"mlr --usage-separator-options\" for information on specifying separators.\n",
       "\n",
       "       Examples:\n",
       "\n",
       "\t mlr nest --explode --values --across-records -f x\n",
       "\t with input record \"x=a;b;c,y=d\" produces output records\n",
       "\t   \"x=a,y=d\"\n",
       "\t   \"x=b,y=d\"\n",
       "\t   \"x=c,y=d\"\n",
       "\t Use --implode to do the reverse.\n",
       "\n",
       "\t mlr nest --explode --values --across-fields -f x\n",
       "\t with input record \"x=a;b;c,y=d\" produces output records\n",
       "\t   \"x_1=a,x_2=b,x_3=c,y=d\"\n",
       "\t Use --implode to do the reverse.\n",
       "\n",
       "\t mlr nest --explode --pairs --across-records -f x\n",
       "\t with input record \"x=a:1;b:2;c:3,y=d\" produces output records\n",
       "\t   \"a=1,y=d\"\n",
       "\t   \"b=2,y=d\"\n",
       "\t   \"c=3,y=d\"\n",
       "\n",
       "\t mlr nest --explode --pairs --across-fields -f x\n",
       "\t with input record \"x=a:1;b:2;c:3,y=d\" produces output records\n",
       "\t   \"a=1,b=2,c=3,y=d\"\n",
       "\n",
       "       Notes:\n",
       "       * With --pairs, --implode doesn't make sense since the original field name has\n",
       "\t been lost.\n",
       "       * The combination \"--implode --values --across-records\" is non-streaming:\n",
       "\t no output records are produced until all input records have been read. In\n",
       "\t particular, this means it won't work in tail -f contexts. But all other flag\n",
       "\t combinations result in streaming (tail -f friendly) data processing.\n",
       "       * It's up to you to ensure that the nested-fs is distinct from your data's IFS:\n",
       "\t e.g. by default the former is semicolon and the latter is comma.\n",
       "       See also mlr reshape.\n",
       "\n",
       "   nothing\n",
       "       Usage: mlr nothing [options]\n",
       "       Drops all input records. Useful for testing, or after tee/print/etc. have\n",
       "       produced other output.\n",
       "\n",
       "   put\n",
       "       Usage: mlr put [options] {expression}\n",
       "       Adds/updates specified field(s). Expressions are semicolon-separated and must\n",
       "       either be assignments, or evaluate to boolean.  Booleans with following\n",
       "       statements in curly braces control whether those statements are executed;\n",
       "       booleans without following curly braces do nothing except side effects (e.g.\n",
       "       regex-captures into \\1, \\2, etc.).\n",
       "\n",
       "       Conversion options:\n",
       "       -S: Keeps field values as strings with no type inference to int or float.\n",
       "       -F: Keeps field values as strings or floats with no inference to int.\n",
       "       All field values are type-inferred to int/float/string unless this behavior is\n",
       "       suppressed with -S or -F.\n",
       "\n",
       "       Output/formatting options:\n",
       "       --oflatsep {string}: Separator to use when flattening multi-level @-variables\n",
       "\t   to output records for emit. Default \":\".\n",
       "       --jknquoteint: For dump output (JSON-formatted), do not quote map keys if non-string.\n",
       "       --jvquoteall: For dump output (JSON-formatted), quote map values even if non-string.\n",
       "       Any of the output-format command-line flags (see mlr -h). Example: using\n",
       "\t mlr --icsv --opprint ... then put --ojson 'tee > \"mytap-\".$a.\".dat\", $*' then ...\n",
       "       the input is CSV, the output is pretty-print tabular, but the tee-file output\n",
       "       is written in JSON format.\n",
       "       --no-fflush: for emit, tee, print, and dump, don't call fflush() after every\n",
       "\t   record.\n",
       "\n",
       "       Expression-specification options:\n",
       "       -f {filename}: the DSL expression is taken from the specified file rather\n",
       "\t   than from the command line. Outer single quotes wrapping the expression\n",
       "\t   should not be placed in the file. If -f is specified more than once,\n",
       "\t   all input files specified using -f are concatenated to produce the expression.\n",
       "\t   (For example, you can define functions in one file and call them from another.)\n",
       "       -e {expression}: You can use this after -f to add an expression. Example use\n",
       "\t   case: define functions/subroutines in a file you specify with -f, then call\n",
       "\t   them with an expression you specify with -e.\n",
       "       (If you mix -e and -f then the expressions are evaluated in the order encountered.\n",
       "       Since the expression pieces are simply concatenated, please be sure to use intervening\n",
       "       semicolons to separate expressions.)\n",
       "\n",
       "       -s name=value: Predefines out-of-stream variable @name to have value \"value\".\n",
       "\t   Thus mlr put put -s foo=97 '$column += @foo' is like\n",
       "\t   mlr put put 'begin {@foo = 97} $column += @foo'.\n",
       "\t   The value part is subject to type-inferencing as specified by -S/-F.\n",
       "\t   May be specified more than once, e.g. -s name1=value1 -s name2=value2.\n",
       "\n",
       "       Tracing options:\n",
       "       -v: Prints the expressions's AST (abstract syntax tree), which gives\n",
       "\t   full transparency on the precedence and associativity rules of\n",
       "\t   Miller's grammar, to stdout.\n",
       "       -a: Prints a low-level stack-allocation trace to stdout.\n",
       "       -t: Prints a low-level parser trace to stderr.\n",
       "       -T: Prints a every statement to stderr as it is executed.\n",
       "\n",
       "       Other options:\n",
       "       -q: Does not include the modified record in the output stream. Useful for when\n",
       "\t   all desired output is in begin and/or end blocks.\n",
       "\n",
       "       Please use a dollar sign for field names and double-quotes for string\n",
       "       literals. If field names have special characters such as \".\" then you might\n",
       "       use braces, e.g. '${field.name}'. Miller built-in variables are\n",
       "       NF NR FNR FILENUM FILENAME M_PI M_E, and ENV[\"namegoeshere\"] to access environment\n",
       "       variables. The environment-variable name may be an expression, e.g. a field\n",
       "       value.\n",
       "\n",
       "       Use # to comment to end of line.\n",
       "\n",
       "       Examples:\n",
       "\t mlr put '$y = log10($x); $z = sqrt($y)'\n",
       "\t mlr put '$x>0.0 { $y=log10($x); $z=sqrt($y) }' # does {...} only if $x > 0.0\n",
       "\t mlr put '$x>0.0;  $y=log10($x); $z=sqrt($y)'\t# does all three statements\n",
       "\t mlr put '$a =~ \"([a-z]+)_([0-9]+);  $b = \"left_\\1\"; $c = \"right_\\2\"'\n",
       "\t mlr put '$a =~ \"([a-z]+)_([0-9]+) { $b = \"left_\\1\"; $c = \"right_\\2\" }'\n",
       "\t mlr put '$filename = FILENAME'\n",
       "\t mlr put '$colored_shape = $color . \"_\" . $shape'\n",
       "\t mlr put '$y = cos($theta); $z = atan2($y, $x)'\n",
       "\t mlr put '$name = sub($name, \"http.*com\"i, \"\")'\n",
       "\t mlr put -q '@sum += $x; end {emit @sum}'\n",
       "\t mlr put -q '@sum[$a] += $x; end {emit @sum, \"a\"}'\n",
       "\t mlr put -q '@sum[$a][$b] += $x; end {emit @sum, \"a\", \"b\"}'\n",
       "\t mlr put -q '@min=min(@min,$x);@max=max(@max,$x); end{emitf @min, @max}'\n",
       "\t mlr put -q 'is_null(@xmax) || $x > @xmax {@xmax=$x; @recmax=$*}; end {emit @recmax}'\n",
       "\t mlr put '\n",
       "\t   $x = 1;\n",
       "\t  #$y = 2;\n",
       "\t   $z = 3\n",
       "\t '\n",
       "\n",
       "       Please see also 'mlr -k' for examples using redirected output.\n",
       "\n",
       "       Please see http://johnkerl.org/miller/doc/reference.html for more information\n",
       "       including function list. Or \"mlr -f\".\n",
       "       Please see in particular:\n",
       "\t http://www.johnkerl.org/miller/doc/reference.html#put\n",
       "\n",
       "   regularize\n",
       "       Usage: mlr regularize\n",
       "       For records seen earlier in the data stream with same field names in\n",
       "       a different order, outputs them with field names in the previously\n",
       "       encountered order.\n",
       "       Example: input records a=1,c=2,b=3, then e=4,d=5, then c=7,a=6,b=8\n",
       "       output as\t      a=1,c=2,b=3, then e=4,d=5, then a=6,c=7,b=8\n",
       "\n",
       "   remove-empty-columns\n",
       "       Usage: mlr remove-empty-columns\n",
       "       Omits fields which are empty on every input row. Non-streaming.\n",
       "\n",
       "   rename\n",
       "       Usage: mlr rename [options] {old1,new1,old2,new2,...}\n",
       "       Renames specified fields.\n",
       "       Options:\n",
       "       -r\t  Treat old field  names as regular expressions. \"ab\", \"a.*b\"\n",
       "\t\t  will match any field name containing the substring \"ab\" or\n",
       "\t\t  matching \"a.*b\", respectively; anchors of the form \"^ab$\",\n",
       "\t\t  \"^a.*b$\" may be used. New field names may be plain strings,\n",
       "\t\t  or may contain capture groups of the form \"\\1\" through\n",
       "\t\t  \"\\9\". Wrapping the regex in double quotes is optional, but\n",
       "\t\t  is required if you wish to follow it with 'i' to indicate\n",
       "\t\t  case-insensitivity.\n",
       "       -g\t  Do global replacement within each field name rather than\n",
       "\t\t  first-match replacement.\n",
       "       Examples:\n",
       "       mlr rename old_name,new_name'\n",
       "       mlr rename old_name_1,new_name_1,old_name_2,new_name_2'\n",
       "       mlr rename -r 'Date_[0-9]+,Date,'  Rename all such fields to be \"Date\"\n",
       "       mlr rename -r '\"Date_[0-9]+\",Date' Same\n",
       "       mlr rename -r 'Date_([0-9]+).*,\\1' Rename all such fields to be of the form 20151015\n",
       "       mlr rename -r '\"name\"i,Name'\t  Rename \"name\", \"Name\", \"NAME\", etc. to \"Name\"\n",
       "\n",
       "   reorder\n",
       "       Usage: mlr reorder [options]\n",
       "       -f {a,b,c}   Field names to reorder.\n",
       "       -e\t    Put specified field names at record end: default is to put\n",
       "\t\t    them at record start.\n",
       "       Examples:\n",
       "       mlr reorder    -f a,b sends input record \"d=4,b=2,a=1,c=3\" to \"a=1,b=2,d=4,c=3\".\n",
       "       mlr reorder -e -f a,b sends input record \"d=4,b=2,a=1,c=3\" to \"d=4,c=3,a=1,b=2\".\n",
       "\n",
       "   repeat\n",
       "       Usage: mlr repeat [options]\n",
       "       Copies input records to output records multiple times.\n",
       "       Options must be exactly one of the following:\n",
       "\t -n {repeat count}  Repeat each input record this many times.\n",
       "\t -f {field name}    Same, but take the repeat count from the specified\n",
       "\t\t\t    field name of each input record.\n",
       "       Example:\n",
       "\t echo x=0 | mlr repeat -n 4 then put '$x=urand()'\n",
       "       produces:\n",
       "\tx=0.488189\n",
       "\tx=0.484973\n",
       "\tx=0.704983\n",
       "\tx=0.147311\n",
       "       Example:\n",
       "\t echo a=1,b=2,c=3 | mlr repeat -f b\n",
       "       produces:\n",
       "\t a=1,b=2,c=3\n",
       "\t a=1,b=2,c=3\n",
       "       Example:\n",
       "\t echo a=1,b=2,c=3 | mlr repeat -f c\n",
       "       produces:\n",
       "\t a=1,b=2,c=3\n",
       "\t a=1,b=2,c=3\n",
       "\t a=1,b=2,c=3\n",
       "\n",
       "   reshape\n",
       "       Usage: mlr reshape [options]\n",
       "       Wide-to-long options:\n",
       "\t -i {input field names}   -o {key-field name,value-field name}\n",
       "\t -r {input field regexes} -o {key-field name,value-field name}\n",
       "\t These pivot/reshape the input data such that the input fields are removed\n",
       "\t and separate records are emitted for each key/value pair.\n",
       "\t Note: this works with tail -f and produces output records for each input\n",
       "\t record seen.\n",
       "       Long-to-wide options:\n",
       "\t -s {key-field name,value-field name}\n",
       "\t These pivot/reshape the input data to undo the wide-to-long operation.\n",
       "\t Note: this does not work with tail -f; it produces output records only after\n",
       "\t all input records have been read.\n",
       "\n",
       "       Examples:\n",
       "\n",
       "\t Input file \"wide.txt\":\n",
       "\t   time       X \t  Y\n",
       "\t   2009-01-01 0.65473572  2.4520609\n",
       "\t   2009-01-02 -0.89248112 0.2154713\n",
       "\t   2009-01-03 0.98012375  1.3179287\n",
       "\n",
       "\t mlr --pprint reshape -i X,Y -o item,value wide.txt\n",
       "\t   time       item value\n",
       "\t   2009-01-01 X    0.65473572\n",
       "\t   2009-01-01 Y    2.4520609\n",
       "\t   2009-01-02 X    -0.89248112\n",
       "\t   2009-01-02 Y    0.2154713\n",
       "\t   2009-01-03 X    0.98012375\n",
       "\t   2009-01-03 Y    1.3179287\n",
       "\n",
       "\t mlr --pprint reshape -r '[A-Z]' -o item,value wide.txt\n",
       "\t   time       item value\n",
       "\t   2009-01-01 X    0.65473572\n",
       "\t   2009-01-01 Y    2.4520609\n",
       "\t   2009-01-02 X    -0.89248112\n",
       "\t   2009-01-02 Y    0.2154713\n",
       "\t   2009-01-03 X    0.98012375\n",
       "\t   2009-01-03 Y    1.3179287\n",
       "\n",
       "\t Input file \"long.txt\":\n",
       "\t   time       item value\n",
       "\t   2009-01-01 X    0.65473572\n",
       "\t   2009-01-01 Y    2.4520609\n",
       "\t   2009-01-02 X    -0.89248112\n",
       "\t   2009-01-02 Y    0.2154713\n",
       "\t   2009-01-03 X    0.98012375\n",
       "\t   2009-01-03 Y    1.3179287\n",
       "\n",
       "\t mlr --pprint reshape -s item,value long.txt\n",
       "\t   time       X \t  Y\n",
       "\t   2009-01-01 0.65473572  2.4520609\n",
       "\t   2009-01-02 -0.89248112 0.2154713\n",
       "\t   2009-01-03 0.98012375  1.3179287\n",
       "       See also mlr nest.\n",
       "\n",
       "   sample\n",
       "       Usage: mlr sample [options]\n",
       "       Reservoir sampling (subsampling without replacement), optionally by category.\n",
       "       -k {count}    Required: number of records to output, total, or by group if using -g.\n",
       "       -g {a,b,c}    Optional: group-by-field names for samples.\n",
       "       See also mlr bootstrap and mlr shuffle.\n",
       "\n",
       "   sec2gmt\n",
       "       Usage: mlr sec2gmt [options] {comma-separated list of field names}\n",
       "       Replaces a numeric field representing seconds since the epoch with the\n",
       "       corresponding GMT timestamp; leaves non-numbers as-is. This is nothing\n",
       "       more than a keystroke-saver for the sec2gmt function:\n",
       "\t mlr sec2gmt time1,time2\n",
       "       is the same as\n",
       "\t mlr put '$time1=sec2gmt($time1);$time2=sec2gmt($time2)'\n",
       "       Options:\n",
       "       -1 through -9: format the seconds using 1..9 decimal places, respectively.\n",
       "\n",
       "   sec2gmtdate\n",
       "       Usage: mlr sec2gmtdate {comma-separated list of field names}\n",
       "       Replaces a numeric field representing seconds since the epoch with the\n",
       "       corresponding GMT year-month-day timestamp; leaves non-numbers as-is.\n",
       "       This is nothing more than a keystroke-saver for the sec2gmtdate function:\n",
       "\t mlr sec2gmtdate time1,time2\n",
       "       is the same as\n",
       "\t mlr put '$time1=sec2gmtdate($time1);$time2=sec2gmtdate($time2)'\n",
       "\n",
       "   seqgen\n",
       "       Usage: mlr seqgen [options]\n",
       "       Produces a sequence of counters.  Discards the input record stream. Produces\n",
       "       output as specified by the following options:\n",
       "       -f {name} Field name for counters; default \"i\".\n",
       "       --start {number} Inclusive start value; default \"1\".\n",
       "       --stop  {number} Inclusive stop value; default \"100\".\n",
       "       --step  {number} Step value; default \"1\".\n",
       "       Start, stop, and/or step may be floating-point. Output is integer if start,\n",
       "       stop, and step are all integers. Step may be negative. It may not be zero\n",
       "       unless start == stop.\n",
       "\n",
       "   shuffle\n",
       "       Usage: mlr shuffle {no options}\n",
       "       Outputs records randomly permuted. No output records are produced until\n",
       "       all input records are read.\n",
       "       See also mlr bootstrap and mlr sample.\n",
       "\n",
       "   skip-trivial-records\n",
       "       Usage: mlr skip-trivial-records [options]\n",
       "       Passes through all records except:\n",
       "       * those with zero fields;\n",
       "       * those for which all fields have empty value.\n",
       "\n",
       "   sort\n",
       "       Usage: mlr sort {flags}\n",
       "       Flags:\n",
       "\t -f  {comma-separated field names}  Lexical ascending\n",
       "\t -n  {comma-separated field names}  Numerical ascending; nulls sort last\n",
       "\t -nf {comma-separated field names}  Same as -n\n",
       "\t -r  {comma-separated field names}  Lexical descending\n",
       "\t -nr {comma-separated field names}  Numerical descending; nulls sort first\n",
       "       Sorts records primarily by the first specified field, secondarily by the second\n",
       "       field, and so on.  (Any records not having all specified sort keys will appear\n",
       "       at the end of the output, in the order they were encountered, regardless of the\n",
       "       specified sort order.) The sort is stable: records that compare equal will sort\n",
       "       in the order they were encountered in the input record stream.\n",
       "\n",
       "       Example:\n",
       "\t mlr sort -f a,b -nr x,y,z\n",
       "       which is the same as:\n",
       "\t mlr sort -f a -f b -nr x -nr y -nr z\n",
       "\n",
       "   stats1\n",
       "       Usage: mlr stats1 [options]\n",
       "       Computes univariate statistics for one or more given fields, accumulated across\n",
       "       the input record stream.\n",
       "       Options:\n",
       "       -a {sum,count,...}  Names of accumulators: p10 p25.2 p50 p98 p100 etc. and/or\n",
       "\t\t\t   one or more of:\n",
       "\t  count     Count instances of fields\n",
       "\t  mode\t    Find most-frequently-occurring values for fields; first-found wins tie\n",
       "\t  antimode  Find least-frequently-occurring values for fields; first-found wins tie\n",
       "\t  sum\t    Compute sums of specified fields\n",
       "\t  mean\t    Compute averages (sample means) of specified fields\n",
       "\t  stddev    Compute sample standard deviation of specified fields\n",
       "\t  var\t    Compute sample variance of specified fields\n",
       "\t  meaneb    Estimate error bars for averages (assuming no sample autocorrelation)\n",
       "\t  skewness  Compute sample skewness of specified fields\n",
       "\t  kurtosis  Compute sample kurtosis of specified fields\n",
       "\t  min\t    Compute minimum values of specified fields\n",
       "\t  max\t    Compute maximum values of specified fields\n",
       "       -f {a,b,c}   Value-field names on which to compute statistics\n",
       "       --fr {regex} Regex for value-field names on which to compute statistics\n",
       "\t\t    (compute statsitics on values in all field names matching regex)\n",
       "       --fx {regex} Inverted regex for value-field names on which to compute statistics\n",
       "\t\t    (compute statsitics on values in all field names not matching regex)\n",
       "       -g {d,e,f}   Optional group-by-field names\n",
       "       --gr {regex} Regex for optional group-by-field names\n",
       "\t\t    (group by values in field names matching regex)\n",
       "       --gx {regex} Inverted regex for optional group-by-field names\n",
       "\t\t    (group by values in field names not matching regex)\n",
       "       --grfx {regex} Shorthand for --gr {regex} --fx {that same regex}\n",
       "       -i\t    Use interpolated percentiles, like R's type=7; default like type=1.\n",
       "\t\t    Not sensical for string-valued fields.\n",
       "       -s\t    Print iterative stats. Useful in tail -f contexts (in which\n",
       "\t\t    case please avoid pprint-format output since end of input\n",
       "\t\t    stream will never be seen).\n",
       "       -F\t    Computes integerable things (e.g. count) in floating point.\n",
       "       Example: mlr stats1 -a min,p10,p50,p90,max -f value -g size,shape\n",
       "       Example: mlr stats1 -a count,mode -f size\n",
       "       Example: mlr stats1 -a count,mode -f size -g shape\n",
       "       Example: mlr stats1 -a count,mode --fr '^[a-h].*$' -gr '^k.*$'\n",
       "\t\tThis computes count and mode statistics on all field names beginning\n",
       "\t\twith a through h, grouped by all field names starting with k.\n",
       "       Notes:\n",
       "       * p50 and median are synonymous.\n",
       "       * min and max output the same results as p0 and p100, respectively, but use\n",
       "\t less memory.\n",
       "       * String-valued data make sense unless arithmetic on them is required,\n",
       "\t e.g. for sum, mean, interpolated percentiles, etc. In case of mixed data,\n",
       "\t numbers are less than strings.\n",
       "       * count and mode allow text input; the rest require numeric input.\n",
       "\t In particular, 1 and 1.0 are distinct text for count and mode.\n",
       "       * When there are mode ties, the first-encountered datum wins.\n",
       "\n",
       "   stats2\n",
       "       Usage: mlr stats2 [options]\n",
       "       Computes bivariate statistics for one or more given field-name pairs,\n",
       "       accumulated across the input record stream.\n",
       "       -a {linreg-ols,corr,...}  Names of accumulators: one or more of:\n",
       "\t linreg-pca   Linear regression using principal component analysis\n",
       "\t linreg-ols   Linear regression using ordinary least squares\n",
       "\t r2\t      Quality metric for linreg-ols (linreg-pca emits its own)\n",
       "\t logireg      Logistic regression\n",
       "\t corr\t      Sample correlation\n",
       "\t cov\t      Sample covariance\n",
       "\t covx\t      Sample-covariance matrix\n",
       "       -f {a,b,c,d}   Value-field name-pairs on which to compute statistics.\n",
       "\t\t      There must be an even number of names.\n",
       "       -g {e,f,g}     Optional group-by-field names.\n",
       "       -v\t      Print additional output for linreg-pca.\n",
       "       -s\t      Print iterative stats. Useful in tail -f contexts (in which\n",
       "\t\t      case please avoid pprint-format output since end of input\n",
       "\t\t      stream will never be seen).\n",
       "       --fit\t      Rather than printing regression parameters, applies them to\n",
       "\t\t      the input data to compute new fit fields. All input records are\n",
       "\t\t      held in memory until end of input stream. Has effect only for\n",
       "\t\t      linreg-ols, linreg-pca, and logireg.\n",
       "       Only one of -s or --fit may be used.\n",
       "       Example: mlr stats2 -a linreg-pca -f x,y\n",
       "       Example: mlr stats2 -a linreg-ols,r2 -f x,y -g size,shape\n",
       "       Example: mlr stats2 -a corr -f x,y\n",
       "\n",
       "   step\n",
       "       Usage: mlr step [options]\n",
       "       Computes values dependent on the previous record, optionally grouped\n",
       "       by category.\n",
       "\n",
       "       Options:\n",
       "       -a {delta,rsum,...}   Names of steppers: comma-separated, one or more of:\n",
       "\t delta\t  Compute differences in field(s) between successive records\n",
       "\t shift\t  Include value(s) in field(s) from previous record, if any\n",
       "\t from-first Compute differences in field(s) from first record\n",
       "\t ratio\t  Compute ratios in field(s) between successive records\n",
       "\t rsum\t  Compute running sums of field(s) between successive records\n",
       "\t counter  Count instances of field(s) between successive records\n",
       "\t ewma\t  Exponentially weighted moving average over successive records\n",
       "       -f {a,b,c} Value-field names on which to compute statistics\n",
       "       -g {d,e,f} Optional group-by-field names\n",
       "       -F\t  Computes integerable things (e.g. counter) in floating point.\n",
       "       -d {x,y,z} Weights for ewma. 1 means current sample gets all weight (no\n",
       "\t\t  smoothing), near under under 1 is light smoothing, near over 0 is\n",
       "\t\t  heavy smoothing. Multiple weights may be specified, e.g.\n",
       "\t\t  \"mlr step -a ewma -f sys_load -d 0.01,0.1,0.9\". Default if omitted\n",
       "\t\t  is \"-d 0.5\".\n",
       "       -o {a,b,c} Custom suffixes for EWMA output fields. If omitted, these default to\n",
       "\t\t  the -d values. If supplied, the number of -o values must be the same\n",
       "\t\t  as the number of -d values.\n",
       "\n",
       "       Examples:\n",
       "\t mlr step -a rsum -f request_size\n",
       "\t mlr step -a delta -f request_size -g hostname\n",
       "\t mlr step -a ewma -d 0.1,0.9 -f x,y\n",
       "\t mlr step -a ewma -d 0.1,0.9 -o smooth,rough -f x,y\n",
       "\t mlr step -a ewma -d 0.1,0.9 -o smooth,rough -f x,y -g group_name\n",
       "\n",
       "       Please see http://johnkerl.org/miller/doc/reference.html#filter or\n",
       "       https://en.wikipedia.org/wiki/Moving_average#Exponential_moving_average\n",
       "       for more information on EWMA.\n",
       "\n",
       "   tac\n",
       "       Usage: mlr tac\n",
       "       Prints records in reverse order from the order in which they were encountered.\n",
       "\n",
       "   tail\n",
       "       Usage: mlr tail [options]\n",
       "       -n {count}    Tail count to print; default 10\n",
       "       -g {a,b,c}    Optional group-by-field names for tail counts\n",
       "       Passes through the last n records, optionally by category.\n",
       "\n",
       "   tee\n",
       "       Usage: mlr tee [options] {filename}\n",
       "       Passes through input records (like mlr cat) but also writes to specified output\n",
       "       file, using output-format flags from the command line (e.g. --ocsv). See also\n",
       "       the \"tee\" keyword within mlr put, which allows data-dependent filenames.\n",
       "       Options:\n",
       "       -a:\t    append to existing file, if any, rather than overwriting.\n",
       "       --no-fflush: don't call fflush() after every record.\n",
       "       Any of the output-format command-line flags (see mlr -h). Example: using\n",
       "\t mlr --icsv --opprint put '...' then tee --ojson ./mytap.dat then stats1 ...\n",
       "       the input is CSV, the output is pretty-print tabular, but the tee-file output\n",
       "       is written in JSON format.\n",
       "\n",
       "   top\n",
       "       Usage: mlr top [options]\n",
       "       -f {a,b,c}    Value-field names for top counts.\n",
       "       -g {d,e,f}    Optional group-by-field names for top counts.\n",
       "       -n {count}    How many records to print per category; default 1.\n",
       "       -a\t     Print all fields for top-value records; default is\n",
       "\t\t     to print only value and group-by fields. Requires a single\n",
       "\t\t     value-field name only.\n",
       "       --min\t     Print top smallest values; default is top largest values.\n",
       "       -F\t     Keep top values as floats even if they look like integers.\n",
       "       -o {name}     Field name for output indices. Default \"top_idx\".\n",
       "       Prints the n records with smallest/largest values at specified fields,\n",
       "       optionally by category.\n",
       "\n",
       "   uniq\n",
       "       Usage: mlr uniq [options]\n",
       "       Prints distinct values for specified field names. With -c, same as\n",
       "       count-distinct. For uniq, -f is a synonym for -g.\n",
       "\n",
       "       Options:\n",
       "       -g {d,e,f}    Group-by-field names for uniq counts.\n",
       "       -c\t     Show repeat counts in addition to unique values.\n",
       "       -n\t     Show only the number of distinct values.\n",
       "       -o {name}     Field name for output count. Default \"count\".\n",
       "       -a\t     Output each unique record only once. Incompatible with -g.\n",
       "\t\t     With -c, produces unique records, with repeat counts for each.\n",
       "\t\t     With -n, produces only one record which is the unique-record count.\n",
       "\t\t     With neither -c nor -n, produces unique records.\n",
       "\n",
       "   unsparsify\n",
       "       Usage: mlr unsparsify [options]\n",
       "       Prints records with the union of field names over all input records.\n",
       "       For field names absent in a given record but present in others, fills in\n",
       "       a value. This verb retains all input before producing any output.\n",
       "\n",
       "       Options:\n",
       "       --fill-with {filler string}  What to fill absent fields with. Defaults to\n",
       "\t\t\t\t    the empty string.\n",
       "\n",
       "       Example: if the input is two records, one being 'a=1,b=2' and the other\n",
       "       being 'b=3,c=4', then the output is the two records 'a=1,b=2,c=' and\n",
       "       'a=,b=3,c=4'.\n",
       "\n",
       "FUNCTIONS FOR FILTER/PUT\n",
       "   +\n",
       "       (class=arithmetic #args=2): Addition.\n",
       "\n",
       "       + (class=arithmetic #args=1): Unary plus.\n",
       "\n",
       "   -\n",
       "       (class=arithmetic #args=2): Subtraction.\n",
       "\n",
       "       - (class=arithmetic #args=1): Unary minus.\n",
       "\n",
       "   *\n",
       "       (class=arithmetic #args=2): Multiplication.\n",
       "\n",
       "   /\n",
       "       (class=arithmetic #args=2): Division.\n",
       "\n",
       "   //\n",
       "       (class=arithmetic #args=2): Integer division: rounds to negative (pythonic).\n",
       "\n",
       "   .+\n",
       "       (class=arithmetic #args=2): Addition, with integer-to-integer overflow\n",
       "\n",
       "       .+ (class=arithmetic #args=1): Unary plus, with integer-to-integer overflow.\n",
       "\n",
       "   .-\n",
       "       (class=arithmetic #args=2): Subtraction, with integer-to-integer overflow.\n",
       "\n",
       "       .- (class=arithmetic #args=1): Unary minus, with integer-to-integer overflow.\n",
       "\n",
       "   .*\n",
       "       (class=arithmetic #args=2): Multiplication, with integer-to-integer overflow.\n",
       "\n",
       "   ./\n",
       "       (class=arithmetic #args=2): Division, with integer-to-integer overflow.\n",
       "\n",
       "   .//\n",
       "       (class=arithmetic #args=2): Integer division: rounds to negative (pythonic), with integer-to-integer overflow.\n",
       "\n",
       "   %\n",
       "       (class=arithmetic #args=2): Remainder; never negative-valued (pythonic).\n",
       "\n",
       "   **\n",
       "       (class=arithmetic #args=2): Exponentiation; same as pow, but as an infix\n",
       "       operator.\n",
       "\n",
       "   |\n",
       "       (class=arithmetic #args=2): Bitwise OR.\n",
       "\n",
       "   ^\n",
       "       (class=arithmetic #args=2): Bitwise XOR.\n",
       "\n",
       "   &\n",
       "       (class=arithmetic #args=2): Bitwise AND.\n",
       "\n",
       "   ~\n",
       "       (class=arithmetic #args=1): Bitwise NOT. Beware '$y=~$x' since =~ is the\n",
       "       regex-match operator: try '$y = ~$x'.\n",
       "\n",
       "   <<\n",
       "       (class=arithmetic #args=2): Bitwise left-shift.\n",
       "\n",
       "   >>\n",
       "       (class=arithmetic #args=2): Bitwise right-shift.\n",
       "\n",
       "   bitcount\n",
       "       (class=arithmetic #args=1): Count of 1-bits\n",
       "\n",
       "   ==\n",
       "       (class=boolean #args=2): String/numeric equality. Mixing number and string\n",
       "       results in string compare.\n",
       "\n",
       "   !=\n",
       "       (class=boolean #args=2): String/numeric inequality. Mixing number and string\n",
       "       results in string compare.\n",
       "\n",
       "   =~\n",
       "       (class=boolean #args=2): String (left-hand side) matches regex (right-hand\n",
       "       side), e.g. '$name =~ \"^a.*b$\"'.\n",
       "\n",
       "   !=~\n",
       "       (class=boolean #args=2): String (left-hand side) does not match regex\n",
       "       (right-hand side), e.g. '$name !=~ \"^a.*b$\"'.\n",
       "\n",
       "   >\n",
       "       (class=boolean #args=2): String/numeric greater-than. Mixing number and string\n",
       "       results in string compare.\n",
       "\n",
       "   >=\n",
       "       (class=boolean #args=2): String/numeric greater-than-or-equals. Mixing number\n",
       "       and string results in string compare.\n",
       "\n",
       "   <\n",
       "       (class=boolean #args=2): String/numeric less-than. Mixing number and string\n",
       "       results in string compare.\n",
       "\n",
       "   <=\n",
       "       (class=boolean #args=2): String/numeric less-than-or-equals. Mixing number\n",
       "       and string results in string compare.\n",
       "\n",
       "   &&\n",
       "       (class=boolean #args=2): Logical AND.\n",
       "\n",
       "   ||\n",
       "       (class=boolean #args=2): Logical OR.\n",
       "\n",
       "   ^^\n",
       "       (class=boolean #args=2): Logical XOR.\n",
       "\n",
       "   !\n",
       "       (class=boolean #args=1): Logical negation.\n",
       "\n",
       "   ? :\n",
       "       (class=boolean #args=3): Ternary operator.\n",
       "\n",
       "   .\n",
       "       (class=string #args=2): String concatenation.\n",
       "\n",
       "   gsub\n",
       "       (class=string #args=3): Example: '$name=gsub($name, \"old\", \"new\")'\n",
       "       (replace all).\n",
       "\n",
       "   regextract\n",
       "       (class=string #args=2): Example: '$name=regextract($name, \"[A-Z]{3}[0-9]{2}\")'\n",
       "       .\n",
       "\n",
       "   regextract_or_else\n",
       "       (class=string #args=3): Example: '$name=regextract_or_else($name, \"[A-Z]{3}[0-9]{2}\", \"default\")'\n",
       "       .\n",
       "\n",
       "   strlen\n",
       "       (class=string #args=1): String length.\n",
       "\n",
       "   sub\n",
       "       (class=string #args=3): Example: '$name=sub($name, \"old\", \"new\")'\n",
       "       (replace once).\n",
       "\n",
       "   ssub\n",
       "       (class=string #args=3): Like sub but does no regexing. No characters are special.\n",
       "\n",
       "   substr\n",
       "       (class=string #args=3): substr(s,m,n) gives substring of s from 0-up position m to n\n",
       "       inclusive. Negative indices -len .. -1 alias to 0 .. len-1.\n",
       "\n",
       "   tolower\n",
       "       (class=string #args=1): Convert string to lowercase.\n",
       "\n",
       "   toupper\n",
       "       (class=string #args=1): Convert string to uppercase.\n",
       "\n",
       "   capitalize\n",
       "       (class=string #args=1): Convert string's first character to uppercase.\n",
       "\n",
       "   lstrip\n",
       "       (class=string #args=1): Strip leading whitespace from string.\n",
       "\n",
       "   rstrip\n",
       "       (class=string #args=1): Strip trailing whitespace from string.\n",
       "\n",
       "   strip\n",
       "       (class=string #args=1): Strip leading and trailing whitespace from string.\n",
       "\n",
       "   collapse_whitespace\n",
       "       (class=string #args=1): Strip repeated whitespace from string.\n",
       "\n",
       "   clean_whitespace\n",
       "       (class=string #args=1): Same as collapse_whitespace and strip.\n",
       "\n",
       "   system\n",
       "       (class=string #args=1): Run command string, yielding its stdout minus final carriage return.\n",
       "\n",
       "   abs\n",
       "       (class=math #args=1): Absolute value.\n",
       "\n",
       "   acos\n",
       "       (class=math #args=1): Inverse trigonometric cosine.\n",
       "\n",
       "   acosh\n",
       "       (class=math #args=1): Inverse hyperbolic cosine.\n",
       "\n",
       "   asin\n",
       "       (class=math #args=1): Inverse trigonometric sine.\n",
       "\n",
       "   asinh\n",
       "       (class=math #args=1): Inverse hyperbolic sine.\n",
       "\n",
       "   atan\n",
       "       (class=math #args=1): One-argument arctangent.\n",
       "\n",
       "   atan2\n",
       "       (class=math #args=2): Two-argument arctangent.\n",
       "\n",
       "   atanh\n",
       "       (class=math #args=1): Inverse hyperbolic tangent.\n",
       "\n",
       "   cbrt\n",
       "       (class=math #args=1): Cube root.\n",
       "\n",
       "   ceil\n",
       "       (class=math #args=1): Ceiling: nearest integer at or above.\n",
       "\n",
       "   cos\n",
       "       (class=math #args=1): Trigonometric cosine.\n",
       "\n",
       "   cosh\n",
       "       (class=math #args=1): Hyperbolic cosine.\n",
       "\n",
       "   erf\n",
       "       (class=math #args=1): Error function.\n",
       "\n",
       "   erfc\n",
       "       (class=math #args=1): Complementary error function.\n",
       "\n",
       "   exp\n",
       "       (class=math #args=1): Exponential function e**x.\n",
       "\n",
       "   expm1\n",
       "       (class=math #args=1): e**x - 1.\n",
       "\n",
       "   floor\n",
       "       (class=math #args=1): Floor: nearest integer at or below.\n",
       "\n",
       "   invqnorm\n",
       "       (class=math #args=1): Inverse of normal cumulative distribution\n",
       "       function. Note that invqorm(urand()) is normally distributed.\n",
       "\n",
       "   log\n",
       "       (class=math #args=1): Natural (base-e) logarithm.\n",
       "\n",
       "   log10\n",
       "       (class=math #args=1): Base-10 logarithm.\n",
       "\n",
       "   log1p\n",
       "       (class=math #args=1): log(1-x).\n",
       "\n",
       "   logifit\n",
       "       (class=math #args=3): Given m and b from logistic regression, compute\n",
       "       fit: $yhat=logifit($x,$m,$b).\n",
       "\n",
       "   madd\n",
       "       (class=math #args=3): a + b mod m (integers)\n",
       "\n",
       "   max\n",
       "       (class=math variadic): max of n numbers; null loses\n",
       "\n",
       "   mexp\n",
       "       (class=math #args=3): a ** b mod m (integers)\n",
       "\n",
       "   min\n",
       "       (class=math variadic): Min of n numbers; null loses\n",
       "\n",
       "   mmul\n",
       "       (class=math #args=3): a * b mod m (integers)\n",
       "\n",
       "   msub\n",
       "       (class=math #args=3): a - b mod m (integers)\n",
       "\n",
       "   pow\n",
       "       (class=math #args=2): Exponentiation; same as **.\n",
       "\n",
       "   qnorm\n",
       "       (class=math #args=1): Normal cumulative distribution function.\n",
       "\n",
       "   round\n",
       "       (class=math #args=1): Round to nearest integer.\n",
       "\n",
       "   roundm\n",
       "       (class=math #args=2): Round to nearest multiple of m: roundm($x,$m) is\n",
       "       the same as round($x/$m)*$m\n",
       "\n",
       "   sgn\n",
       "       (class=math #args=1): +1 for positive input, 0 for zero input, -1 for\n",
       "       negative input.\n",
       "\n",
       "   sin\n",
       "       (class=math #args=1): Trigonometric sine.\n",
       "\n",
       "   sinh\n",
       "       (class=math #args=1): Hyperbolic sine.\n",
       "\n",
       "   sqrt\n",
       "       (class=math #args=1): Square root.\n",
       "\n",
       "   tan\n",
       "       (class=math #args=1): Trigonometric tangent.\n",
       "\n",
       "   tanh\n",
       "       (class=math #args=1): Hyperbolic tangent.\n",
       "\n",
       "   urand\n",
       "       (class=math #args=0): Floating-point numbers uniformly distributed on the unit interval.\n",
       "       Int-valued example: '$n=floor(20+urand()*11)'.\n",
       "\n",
       "   urandrange\n",
       "       (class=math #args=2): Floating-point numbers uniformly distributed on the interval [a, b).\n",
       "\n",
       "   urand32\n",
       "       (class=math #args=0): Integer uniformly distributed 0 and 2**32-1\n",
       "       inclusive.\n",
       "\n",
       "   urandint\n",
       "       (class=math #args=2): Integer uniformly distributed between inclusive\n",
       "       integer endpoints.\n",
       "\n",
       "   dhms2fsec\n",
       "       (class=time #args=1): Recovers floating-point seconds as in\n",
       "       dhms2fsec(\"5d18h53m20.250000s\") = 500000.250000\n",
       "\n",
       "   dhms2sec\n",
       "       (class=time #args=1): Recovers integer seconds as in\n",
       "       dhms2sec(\"5d18h53m20s\") = 500000\n",
       "\n",
       "   fsec2dhms\n",
       "       (class=time #args=1): Formats floating-point seconds as in\n",
       "       fsec2dhms(500000.25) = \"5d18h53m20.250000s\"\n",
       "\n",
       "   fsec2hms\n",
       "       (class=time #args=1): Formats floating-point seconds as in\n",
       "       fsec2hms(5000.25) = \"01:23:20.250000\"\n",
       "\n",
       "   gmt2sec\n",
       "       (class=time #args=1): Parses GMT timestamp as integer seconds since\n",
       "       the epoch.\n",
       "\n",
       "   localtime2sec\n",
       "       (class=time #args=1): Parses local timestamp as integer seconds since\n",
       "       the epoch. Consults $TZ environment variable.\n",
       "\n",
       "   hms2fsec\n",
       "       (class=time #args=1): Recovers floating-point seconds as in\n",
       "       hms2fsec(\"01:23:20.250000\") = 5000.250000\n",
       "\n",
       "   hms2sec\n",
       "       (class=time #args=1): Recovers integer seconds as in\n",
       "       hms2sec(\"01:23:20\") = 5000\n",
       "\n",
       "   sec2dhms\n",
       "       (class=time #args=1): Formats integer seconds as in sec2dhms(500000)\n",
       "       = \"5d18h53m20s\"\n",
       "\n",
       "   sec2gmt\n",
       "       (class=time #args=1): Formats seconds since epoch (integer part)\n",
       "       as GMT timestamp, e.g. sec2gmt(1440768801.7) = \"2015-08-28T13:33:21Z\".\n",
       "       Leaves non-numbers as-is.\n",
       "\n",
       "       sec2gmt (class=time #args=2): Formats seconds since epoch as GMT timestamp with n\n",
       "       decimal places for seconds, e.g. sec2gmt(1440768801.7,1) = \"2015-08-28T13:33:21.7Z\".\n",
       "       Leaves non-numbers as-is.\n",
       "\n",
       "   sec2gmtdate\n",
       "       (class=time #args=1): Formats seconds since epoch (integer part)\n",
       "       as GMT timestamp with year-month-date, e.g. sec2gmtdate(1440768801.7) = \"2015-08-28\".\n",
       "       Leaves non-numbers as-is.\n",
       "\n",
       "   sec2localtime\n",
       "       (class=time #args=1): Formats seconds since epoch (integer part)\n",
       "       as local timestamp, e.g. sec2localtime(1440768801.7) = \"2015-08-28T13:33:21Z\".\n",
       "       Consults $TZ environment variable. Leaves non-numbers as-is.\n",
       "\n",
       "       sec2localtime (class=time #args=2): Formats seconds since epoch as local timestamp with n\n",
       "       decimal places for seconds, e.g. sec2localtime(1440768801.7,1) = \"2015-08-28T13:33:21.7Z\".\n",
       "       Consults $TZ environment variable. Leaves non-numbers as-is.\n",
       "\n",
       "   sec2localdate\n",
       "       (class=time #args=1): Formats seconds since epoch (integer part)\n",
       "       as local timestamp with year-month-date, e.g. sec2localdate(1440768801.7) = \"2015-08-28\".\n",
       "       Consults $TZ environment variable. Leaves non-numbers as-is.\n",
       "\n",
       "   sec2hms\n",
       "       (class=time #args=1): Formats integer seconds as in\n",
       "       sec2hms(5000) = \"01:23:20\"\n",
       "\n",
       "   strftime\n",
       "       (class=time #args=2): Formats seconds since the epoch as timestamp, e.g.\n",
       "       strftime(1440768801.7,\"%Y-%m-%dT%H:%M:%SZ\") = \"2015-08-28T13:33:21Z\", and\n",
       "       strftime(1440768801.7,\"%Y-%m-%dT%H:%M:%3SZ\") = \"2015-08-28T13:33:21.700Z\".\n",
       "       Format strings are as in the C library (please see \"man strftime\" on your system),\n",
       "       with the Miller-specific addition of \"%1S\" through \"%9S\" which format the seconds\n",
       "       with 1 through 9 decimal places, respectively. (\"%S\" uses no decimal places.)\n",
       "       See also strftime_local.\n",
       "\n",
       "   strftime_local\n",
       "       (class=time #args=2): Like strftime but consults the $TZ environment variable to get local time zone.\n",
       "\n",
       "   strptime\n",
       "       (class=time #args=2): Parses timestamp as floating-point seconds since the epoch,\n",
       "       e.g. strptime(\"2015-08-28T13:33:21Z\",\"%Y-%m-%dT%H:%M:%SZ\") = 1440768801.000000,\n",
       "       and  strptime(\"2015-08-28T13:33:21.345Z\",\"%Y-%m-%dT%H:%M:%SZ\") = 1440768801.345000.\n",
       "       See also strptime_local.\n",
       "\n",
       "   strptime_local\n",
       "       (class=time #args=2): Like strptime, but consults $TZ environment variable to find and use local timezone.\n",
       "\n",
       "   systime\n",
       "       (class=time #args=0): Floating-point seconds since the epoch,\n",
       "       e.g. 1440768801.748936.\n",
       "\n",
       "   is_absent\n",
       "       (class=typing #args=1): False if field is present in input, true otherwise\n",
       "\n",
       "   is_bool\n",
       "       (class=typing #args=1): True if field is present with boolean value. Synonymous with is_boolean.\n",
       "\n",
       "   is_boolean\n",
       "       (class=typing #args=1): True if field is present with boolean value. Synonymous with is_bool.\n",
       "\n",
       "   is_empty\n",
       "       (class=typing #args=1): True if field is present in input with empty string value, false otherwise.\n",
       "\n",
       "   is_empty_map\n",
       "       (class=typing #args=1): True if argument is a map which is empty.\n",
       "\n",
       "   is_float\n",
       "       (class=typing #args=1): True if field is present with value inferred to be float\n",
       "\n",
       "   is_int\n",
       "       (class=typing #args=1): True if field is present with value inferred to be int\n",
       "\n",
       "   is_map\n",
       "       (class=typing #args=1): True if argument is a map.\n",
       "\n",
       "   is_nonempty_map\n",
       "       (class=typing #args=1): True if argument is a map which is non-empty.\n",
       "\n",
       "   is_not_empty\n",
       "       (class=typing #args=1): False if field is present in input with empty value, true otherwise\n",
       "\n",
       "   is_not_map\n",
       "       (class=typing #args=1): True if argument is not a map.\n",
       "\n",
       "   is_not_null\n",
       "       (class=typing #args=1): False if argument is null (empty or absent), true otherwise.\n",
       "\n",
       "   is_null\n",
       "       (class=typing #args=1): True if argument is null (empty or absent), false otherwise.\n",
       "\n",
       "   is_numeric\n",
       "       (class=typing #args=1): True if field is present with value inferred to be int or float\n",
       "\n",
       "   is_present\n",
       "       (class=typing #args=1): True if field is present in input, false otherwise.\n",
       "\n",
       "   is_string\n",
       "       (class=typing #args=1): True if field is present with string (including empty-string) value\n",
       "\n",
       "   asserting_absent\n",
       "       (class=typing #args=1): Returns argument if it is absent in the input data, else\n",
       "       throws an error.\n",
       "\n",
       "   asserting_bool\n",
       "       (class=typing #args=1): Returns argument if it is present with boolean value, else\n",
       "       throws an error.\n",
       "\n",
       "   asserting_boolean\n",
       "       (class=typing #args=1): Returns argument if it is present with boolean value, else\n",
       "       throws an error.\n",
       "\n",
       "   asserting_empty\n",
       "       (class=typing #args=1): Returns argument if it is present in input with empty value,\n",
       "       else throws an error.\n",
       "\n",
       "   asserting_empty_map\n",
       "       (class=typing #args=1): Returns argument if it is a map with empty value, else\n",
       "       throws an error.\n",
       "\n",
       "   asserting_float\n",
       "       (class=typing #args=1): Returns argument if it is present with float value, else\n",
       "       throws an error.\n",
       "\n",
       "   asserting_int\n",
       "       (class=typing #args=1): Returns argument if it is present with int value, else\n",
       "       throws an error.\n",
       "\n",
       "   asserting_map\n",
       "       (class=typing #args=1): Returns argument if it is a map, else throws an error.\n",
       "\n",
       "   asserting_nonempty_map\n",
       "       (class=typing #args=1): Returns argument if it is a non-empty map, else throws\n",
       "       an error.\n",
       "\n",
       "   asserting_not_empty\n",
       "       (class=typing #args=1): Returns argument if it is present in input with non-empty\n",
       "       value, else throws an error.\n",
       "\n",
       "   asserting_not_map\n",
       "       (class=typing #args=1): Returns argument if it is not a map, else throws an error.\n",
       "\n",
       "   asserting_not_null\n",
       "       (class=typing #args=1): Returns argument if it is non-null (non-empty and non-absent),\n",
       "       else throws an error.\n",
       "\n",
       "   asserting_null\n",
       "       (class=typing #args=1): Returns argument if it is null (empty or absent), else throws\n",
       "       an error.\n",
       "\n",
       "   asserting_numeric\n",
       "       (class=typing #args=1): Returns argument if it is present with int or float value,\n",
       "       else throws an error.\n",
       "\n",
       "   asserting_present\n",
       "       (class=typing #args=1): Returns argument if it is present in input, else throws\n",
       "       an error.\n",
       "\n",
       "   asserting_string\n",
       "       (class=typing #args=1): Returns argument if it is present with string (including\n",
       "       empty-string) value, else throws an error.\n",
       "\n",
       "   boolean\n",
       "       (class=conversion #args=1): Convert int/float/bool/string to boolean.\n",
       "\n",
       "   float\n",
       "       (class=conversion #args=1): Convert int/float/bool/string to float.\n",
       "\n",
       "   fmtnum\n",
       "       (class=conversion #args=2): Convert int/float/bool to string using\n",
       "       printf-style format string, e.g. '$s = fmtnum($n, \"%06lld\")'. WARNING: Miller numbers\n",
       "       are all long long or double. If you use formats like %d or %f, behavior is undefined.\n",
       "\n",
       "   hexfmt\n",
       "       (class=conversion #args=1): Convert int to string, e.g. 255 to \"0xff\".\n",
       "\n",
       "   int\n",
       "       (class=conversion #args=1): Convert int/float/bool/string to int.\n",
       "\n",
       "   string\n",
       "       (class=conversion #args=1): Convert int/float/bool/string to string.\n",
       "\n",
       "   typeof\n",
       "       (class=conversion #args=1): Convert argument to type of argument (e.g.\n",
       "       MT_STRING). For debug.\n",
       "\n",
       "   depth\n",
       "       (class=maps #args=1): Prints maximum depth of hashmap: ''. Scalars have depth 0.\n",
       "\n",
       "   haskey\n",
       "       (class=maps #args=2): True/false if map has/hasn't key, e.g. 'haskey($*, \"a\")' or\n",
       "       'haskey(mymap, mykey)'. Error if 1st argument is not a map.\n",
       "\n",
       "   joink\n",
       "       (class=maps #args=2): Makes string from map keys. E.g. 'joink($*, \",\")'.\n",
       "\n",
       "   joinkv\n",
       "       (class=maps #args=3): Makes string from map key-value pairs. E.g. 'joinkv(@v[2], \"=\", \",\")'\n",
       "\n",
       "   joinv\n",
       "       (class=maps #args=2): Makes string from map keys. E.g. 'joinv(mymap, \",\")'.\n",
       "\n",
       "   leafcount\n",
       "       (class=maps #args=1): Counts total number of terminal values in hashmap. For single-level maps,\n",
       "       same as length.\n",
       "\n",
       "   length\n",
       "       (class=maps #args=1): Counts number of top-level entries in hashmap. Scalars have length 1.\n",
       "\n",
       "   mapdiff\n",
       "       (class=maps variadic): With 0 args, returns empty map. With 1 arg, returns copy of arg.\n",
       "       With 2 or more, returns copy of arg 1 with all keys from any of remaining argument maps removed.\n",
       "\n",
       "   mapexcept\n",
       "       (class=maps variadic): Returns a map with keys from remaining arguments, if any, unset.\n",
       "       E.g. 'mapexcept({1:2,3:4,5:6}, 1, 5, 7)' is '{3:4}'.\n",
       "\n",
       "   mapselect\n",
       "       (class=maps variadic): Returns a map with only keys from remaining arguments set.\n",
       "       E.g. 'mapselect({1:2,3:4,5:6}, 1, 5, 7)' is '{1:2,5:6}'.\n",
       "\n",
       "   mapsum\n",
       "       (class=maps variadic): With 0 args, returns empty map. With >= 1 arg, returns a map with\n",
       "       key-value pairs from all arguments. Rightmost collisions win, e.g. 'mapsum({1:2,3:4},{1:5})' is '{1:5,3:4}'.\n",
       "\n",
       "   splitkv\n",
       "       (class=maps #args=3): Splits string by separators into map with type inference.\n",
       "       E.g. 'splitkv(\"a=1,b=2,c=3\", \"=\", \",\")' gives '{\"a\" : 1, \"b\" : 2, \"c\" : 3}'.\n",
       "\n",
       "   splitkvx\n",
       "       (class=maps #args=3): Splits string by separators into map without type inference (keys and\n",
       "       values are strings). E.g. 'splitkv(\"a=1,b=2,c=3\", \"=\", \",\")' gives\n",
       "       '{\"a\" : \"1\", \"b\" : \"2\", \"c\" : \"3\"}'.\n",
       "\n",
       "   splitnv\n",
       "       (class=maps #args=2): Splits string by separator into integer-indexed map with type inference.\n",
       "       E.g. 'splitnv(\"a,b,c\" , \",\")' gives '{1 : \"a\", 2 : \"b\", 3 : \"c\"}'.\n",
       "\n",
       "   splitnvx\n",
       "       (class=maps #args=2): Splits string by separator into integer-indexed map without type\n",
       "       inference (values are strings). E.g. 'splitnv(\"4,5,6\" , \",\")' gives '{1 : \"4\", 2 : \"5\", 3 : \"6\"}'.\n",
       "\n",
       "KEYWORDS FOR PUT AND FILTER\n",
       "   all\n",
       "       all: used in \"emit\", \"emitp\", and \"unset\" as a synonym for @*\n",
       "\n",
       "   begin\n",
       "       begin: defines a block of statements to be executed before input records\n",
       "       are ingested. The body statements must be wrapped in curly braces.\n",
       "       Example: 'begin { @count = 0 }'\n",
       "\n",
       "   bool\n",
       "       bool: declares a boolean local variable in the current curly-braced scope.\n",
       "       Type-checking happens at assignment: 'bool b = 1' is an error.\n",
       "\n",
       "   break\n",
       "       break: causes execution to continue after the body of the current\n",
       "       for/while/do-while loop.\n",
       "\n",
       "   call\n",
       "       call: used for invoking a user-defined subroutine.\n",
       "       Example: 'subr s(k,v) { print k . \" is \" . v} call s(\"a\", $a)'\n",
       "\n",
       "   continue\n",
       "       continue: causes execution to skip the remaining statements in the body of\n",
       "       the current for/while/do-while loop. For-loop increments are still applied.\n",
       "\n",
       "   do\n",
       "       do: with \"while\", introduces a do-while loop. The body statements must be wrapped\n",
       "       in curly braces.\n",
       "\n",
       "   dump\n",
       "       dump: prints all currently defined out-of-stream variables immediately\n",
       "\t to stdout as JSON.\n",
       "\n",
       "\t With >, >>, or |, the data do not become part of the output record stream but\n",
       "\t are instead redirected.\n",
       "\n",
       "\t The > and >> are for write and append, as in the shell, but (as with awk) the\n",
       "\t file-overwrite for > is on first write, not per record. The | is for piping to\n",
       "\t a process which will process the data. There will be one open file for each\n",
       "\t distinct file name (for > and >>) or one subordinate process for each distinct\n",
       "\t value of the piped-to command (for |). Output-formatting flags are taken from\n",
       "\t the main command line.\n",
       "\n",
       "\t Example: mlr --from f.dat put -q '@v[NR]=$*; end { dump }'\n",
       "\t Example: mlr --from f.dat put -q '@v[NR]=$*; end { dump >  \"mytap.dat\"}'\n",
       "\t Example: mlr --from f.dat put -q '@v[NR]=$*; end { dump >> \"mytap.dat\"}'\n",
       "\t Example: mlr --from f.dat put -q '@v[NR]=$*; end { dump | \"jq .[]\"}'\n",
       "\n",
       "   edump\n",
       "       edump: prints all currently defined out-of-stream variables immediately\n",
       "\t to stderr as JSON.\n",
       "\n",
       "\t Example: mlr --from f.dat put -q '@v[NR]=$*; end { edump }'\n",
       "\n",
       "   elif\n",
       "       elif: the way Miller spells \"else if\". The body statements must be wrapped\n",
       "       in curly braces.\n",
       "\n",
       "   else\n",
       "       else: terminates an if/elif/elif chain. The body statements must be wrapped\n",
       "       in curly braces.\n",
       "\n",
       "   emit\n",
       "       emit: inserts an out-of-stream variable into the output record stream. Hashmap\n",
       "\t indices present in the data but not slotted by emit arguments are not output.\n",
       "\n",
       "\t With >, >>, or |, the data do not become part of the output record stream but\n",
       "\t are instead redirected.\n",
       "\n",
       "\t The > and >> are for write and append, as in the shell, but (as with awk) the\n",
       "\t file-overwrite for > is on first write, not per record. The | is for piping to\n",
       "\t a process which will process the data. There will be one open file for each\n",
       "\t distinct file name (for > and >>) or one subordinate process for each distinct\n",
       "\t value of the piped-to command (for |). Output-formatting flags are taken from\n",
       "\t the main command line.\n",
       "\n",
       "\t You can use any of the output-format command-line flags, e.g. --ocsv, --ofs,\n",
       "\t etc., to control the format of the output if the output is redirected. See also mlr -h.\n",
       "\n",
       "\t Example: mlr --from f.dat put 'emit >\t\"/tmp/data-\".$a, $*'\n",
       "\t Example: mlr --from f.dat put 'emit >\t\"/tmp/data-\".$a, mapexcept($*, \"a\")'\n",
       "\t Example: mlr --from f.dat put '@sums[$a][$b]+=$x; emit @sums'\n",
       "\t Example: mlr --from f.dat put --ojson '@sums[$a][$b]+=$x; emit > \"tap-\".$a.$b.\".dat\", @sums'\n",
       "\t Example: mlr --from f.dat put '@sums[$a][$b]+=$x; emit @sums, \"index1\", \"index2\"'\n",
       "\t Example: mlr --from f.dat put '@sums[$a][$b]+=$x; emit @*, \"index1\", \"index2\"'\n",
       "\t Example: mlr --from f.dat put '@sums[$a][$b]+=$x; emit >  \"mytap.dat\", @*, \"index1\", \"index2\"'\n",
       "\t Example: mlr --from f.dat put '@sums[$a][$b]+=$x; emit >> \"mytap.dat\", @*, \"index1\", \"index2\"'\n",
       "\t Example: mlr --from f.dat put '@sums[$a][$b]+=$x; emit | \"gzip > mytap.dat.gz\", @*, \"index1\", \"index2\"'\n",
       "\t Example: mlr --from f.dat put '@sums[$a][$b]+=$x; emit > stderr, @*, \"index1\", \"index2\"'\n",
       "\t Example: mlr --from f.dat put '@sums[$a][$b]+=$x; emit | \"grep somepattern\", @*, \"index1\", \"index2\"'\n",
       "\n",
       "\t Please see http://johnkerl.org/miller/doc for more information.\n",
       "\n",
       "   emitf\n",
       "       emitf: inserts non-indexed out-of-stream variable(s) side-by-side into the\n",
       "\t output record stream.\n",
       "\n",
       "\t With >, >>, or |, the data do not become part of the output record stream but\n",
       "\t are instead redirected.\n",
       "\n",
       "\t The > and >> are for write and append, as in the shell, but (as with awk) the\n",
       "\t file-overwrite for > is on first write, not per record. The | is for piping to\n",
       "\t a process which will process the data. There will be one open file for each\n",
       "\t distinct file name (for > and >>) or one subordinate process for each distinct\n",
       "\t value of the piped-to command (for |). Output-formatting flags are taken from\n",
       "\t the main command line.\n",
       "\n",
       "\t You can use any of the output-format command-line flags, e.g. --ocsv, --ofs,\n",
       "\t etc., to control the format of the output if the output is redirected. See also mlr -h.\n",
       "\n",
       "\t Example: mlr --from f.dat put '@a=$i;@b+=$x;@c+=$y; emitf @a'\n",
       "\t Example: mlr --from f.dat put --oxtab '@a=$i;@b+=$x;@c+=$y; emitf > \"tap-\".$i.\".dat\", @a'\n",
       "\t Example: mlr --from f.dat put '@a=$i;@b+=$x;@c+=$y; emitf @a, @b, @c'\n",
       "\t Example: mlr --from f.dat put '@a=$i;@b+=$x;@c+=$y; emitf > \"mytap.dat\", @a, @b, @c'\n",
       "\t Example: mlr --from f.dat put '@a=$i;@b+=$x;@c+=$y; emitf >> \"mytap.dat\", @a, @b, @c'\n",
       "\t Example: mlr --from f.dat put '@a=$i;@b+=$x;@c+=$y; emitf > stderr, @a, @b, @c'\n",
       "\t Example: mlr --from f.dat put '@a=$i;@b+=$x;@c+=$y; emitf | \"grep somepattern\", @a, @b, @c'\n",
       "\t Example: mlr --from f.dat put '@a=$i;@b+=$x;@c+=$y; emitf | \"grep somepattern > mytap.dat\", @a, @b, @c'\n",
       "\n",
       "\t Please see http://johnkerl.org/miller/doc for more information.\n",
       "\n",
       "   emitp\n",
       "       emitp: inserts an out-of-stream variable into the output record stream.\n",
       "\t Hashmap indices present in the data but not slotted by emitp arguments are\n",
       "\t output concatenated with \":\".\n",
       "\n",
       "\t With >, >>, or |, the data do not become part of the output record stream but\n",
       "\t are instead redirected.\n",
       "\n",
       "\t The > and >> are for write and append, as in the shell, but (as with awk) the\n",
       "\t file-overwrite for > is on first write, not per record. The | is for piping to\n",
       "\t a process which will process the data. There will be one open file for each\n",
       "\t distinct file name (for > and >>) or one subordinate process for each distinct\n",
       "\t value of the piped-to command (for |). Output-formatting flags are taken from\n",
       "\t the main command line.\n",
       "\n",
       "\t You can use any of the output-format command-line flags, e.g. --ocsv, --ofs,\n",
       "\t etc., to control the format of the output if the output is redirected. See also mlr -h.\n",
       "\n",
       "\t Example: mlr --from f.dat put '@sums[$a][$b]+=$x; emitp @sums'\n",
       "\t Example: mlr --from f.dat put --opprint '@sums[$a][$b]+=$x; emitp > \"tap-\".$a.$b.\".dat\", @sums'\n",
       "\t Example: mlr --from f.dat put '@sums[$a][$b]+=$x; emitp @sums, \"index1\", \"index2\"'\n",
       "\t Example: mlr --from f.dat put '@sums[$a][$b]+=$x; emitp @*, \"index1\", \"index2\"'\n",
       "\t Example: mlr --from f.dat put '@sums[$a][$b]+=$x; emitp >  \"mytap.dat\", @*, \"index1\", \"index2\"'\n",
       "\t Example: mlr --from f.dat put '@sums[$a][$b]+=$x; emitp >> \"mytap.dat\", @*, \"index1\", \"index2\"'\n",
       "\t Example: mlr --from f.dat put '@sums[$a][$b]+=$x; emitp | \"gzip > mytap.dat.gz\", @*, \"index1\", \"index2\"'\n",
       "\t Example: mlr --from f.dat put '@sums[$a][$b]+=$x; emitp > stderr, @*, \"index1\", \"index2\"'\n",
       "\t Example: mlr --from f.dat put '@sums[$a][$b]+=$x; emitp | \"grep somepattern\", @*, \"index1\", \"index2\"'\n",
       "\n",
       "\t Please see http://johnkerl.org/miller/doc for more information.\n",
       "\n",
       "   end\n",
       "       end: defines a block of statements to be executed after input records\n",
       "       are ingested. The body statements must be wrapped in curly braces.\n",
       "       Example: 'end { emit @count }'\n",
       "       Example: 'end { eprint \"Final count is \" . @count }'\n",
       "\n",
       "   eprint\n",
       "       eprint: prints expression immediately to stderr.\n",
       "\t Example: mlr --from f.dat put -q 'eprint \"The sum of x and y is \".($x+$y)'\n",
       "\t Example: mlr --from f.dat put -q 'for (k, v in $*) { eprint k . \" => \" . v }'\n",
       "\t Example: mlr --from f.dat put\t'(NR % 1000 == 0) { eprint \"Checkpoint \".NR}'\n",
       "\n",
       "   eprintn\n",
       "       eprintn: prints expression immediately to stderr, without trailing newline.\n",
       "\t Example: mlr --from f.dat put -q 'eprintn \"The sum of x and y is \".($x+$y); eprint \"\"'\n",
       "\n",
       "   false\n",
       "       false: the boolean literal value.\n",
       "\n",
       "   filter\n",
       "       filter: includes/excludes the record in the output record stream.\n",
       "\n",
       "\t Example: mlr --from f.dat put 'filter (NR == 2 || $x > 5.4)'\n",
       "\n",
       "\t Instead of put with 'filter false' you can simply use put -q.\tThe following\n",
       "\t uses the input record to accumulate data but only prints the running sum\n",
       "\t without printing the input record:\n",
       "\n",
       "\t Example: mlr --from f.dat put -q '@running_sum += $x * $y; emit @running_sum'\n",
       "\n",
       "   float\n",
       "       float: declares a floating-point local variable in the current curly-braced scope.\n",
       "       Type-checking happens at assignment: 'float x = 0' is an error.\n",
       "\n",
       "   for\n",
       "       for: defines a for-loop using one of three styles. The body statements must\n",
       "       be wrapped in curly braces.\n",
       "       For-loop over stream record:\n",
       "\t Example:  'for (k, v in $*) { ... }'\n",
       "       For-loop over out-of-stream variables:\n",
       "\t Example: 'for (k, v in @counts) { ... }'\n",
       "\t Example: 'for ((k1, k2), v in @counts) { ... }'\n",
       "\t Example: 'for ((k1, k2, k3), v in @*) { ... }'\n",
       "       C-style for-loop:\n",
       "\t Example:  'for (var i = 0, var b = 1; i < 10; i += 1, b *= 2) { ... }'\n",
       "\n",
       "   func\n",
       "       func: used for defining a user-defined function.\n",
       "       Example: 'func f(a,b) { return sqrt(a**2+b**2)} $d = f($x, $y)'\n",
       "\n",
       "   if\n",
       "       if: starts an if/elif/elif chain. The body statements must be wrapped\n",
       "       in curly braces.\n",
       "\n",
       "   in\n",
       "       in: used in for-loops over stream records or out-of-stream variables.\n",
       "\n",
       "   int\n",
       "       int: declares an integer local variable in the current curly-braced scope.\n",
       "       Type-checking happens at assignment: 'int x = 0.0' is an error.\n",
       "\n",
       "   map\n",
       "       map: declares an map-valued local variable in the current curly-braced scope.\n",
       "       Type-checking happens at assignment: 'map b = 0' is an error. map b = {} is\n",
       "       always OK. map b = a is OK or not depending on whether a is a map.\n",
       "\n",
       "   num\n",
       "       num: declares an int/float local variable in the current curly-braced scope.\n",
       "       Type-checking happens at assignment: 'num b = true' is an error.\n",
       "\n",
       "   print\n",
       "       print: prints expression immediately to stdout.\n",
       "\t Example: mlr --from f.dat put -q 'print \"The sum of x and y is \".($x+$y)'\n",
       "\t Example: mlr --from f.dat put -q 'for (k, v in $*) { print k . \" => \" . v }'\n",
       "\t Example: mlr --from f.dat put\t'(NR % 1000 == 0) { print > stderr, \"Checkpoint \".NR}'\n",
       "\n",
       "   printn\n",
       "       printn: prints expression immediately to stdout, without trailing newline.\n",
       "\t Example: mlr --from f.dat put -q 'printn \".\"; end { print \"\" }'\n",
       "\n",
       "   return\n",
       "       return: specifies the return value from a user-defined function.\n",
       "       Omitted return statements (including via if-branches) result in an absent-null\n",
       "       return value, which in turns results in a skipped assignment to an LHS.\n",
       "\n",
       "   stderr\n",
       "       stderr: Used for tee, emit, emitf, emitp, print, and dump in place of filename\n",
       "\t to print to standard error.\n",
       "\n",
       "   stdout\n",
       "       stdout: Used for tee, emit, emitf, emitp, print, and dump in place of filename\n",
       "\t to print to standard output.\n",
       "\n",
       "   str\n",
       "       str: declares a string local variable in the current curly-braced scope.\n",
       "       Type-checking happens at assignment.\n",
       "\n",
       "   subr\n",
       "       subr: used for defining a subroutine.\n",
       "       Example: 'subr s(k,v) { print k . \" is \" . v} call s(\"a\", $a)'\n",
       "\n",
       "   tee\n",
       "       tee: prints the current record to specified file.\n",
       "\t This is an immediate print to the specified file (except for pprint format\n",
       "\t which of course waits until the end of the input stream to format all output).\n",
       "\n",
       "\t The > and >> are for write and append, as in the shell, but (as with awk) the\n",
       "\t file-overwrite for > is on first write, not per record. The | is for piping to\n",
       "\t a process which will process the data. There will be one open file for each\n",
       "\t distinct file name (for > and >>) or one subordinate process for each distinct\n",
       "\t value of the piped-to command (for |). Output-formatting flags are taken from\n",
       "\t the main command line.\n",
       "\n",
       "\t You can use any of the output-format command-line flags, e.g. --ocsv, --ofs,\n",
       "\t etc., to control the format of the output. See also mlr -h.\n",
       "\n",
       "\t emit with redirect and tee with redirect are identical, except tee can only\n",
       "\t output $*.\n",
       "\n",
       "\t Example: mlr --from f.dat put 'tee >  \"/tmp/data-\".$a, $*'\n",
       "\t Example: mlr --from f.dat put 'tee >> \"/tmp/data-\".$a.$b, $*'\n",
       "\t Example: mlr --from f.dat put 'tee >  stderr, $*'\n",
       "\t Example: mlr --from f.dat put -q 'tee | \"tr [a-z\\] [A-Z\\]\", $*'\n",
       "\t Example: mlr --from f.dat put -q 'tee | \"tr [a-z\\] [A-Z\\] > /tmp/data-\".$a, $*'\n",
       "\t Example: mlr --from f.dat put -q 'tee | \"gzip > /tmp/data-\".$a.\".gz\", $*'\n",
       "\t Example: mlr --from f.dat put -q --ojson 'tee | \"gzip > /tmp/data-\".$a.\".gz\", $*'\n",
       "\n",
       "   true\n",
       "       true: the boolean literal value.\n",
       "\n",
       "   unset\n",
       "       unset: clears field(s) from the current record, or an out-of-stream or local variable.\n",
       "\n",
       "\t Example: mlr --from f.dat put 'unset $x'\n",
       "\t Example: mlr --from f.dat put 'unset $*'\n",
       "\t Example: mlr --from f.dat put 'for (k, v in $*) { if (k =~ \"a.*\") { unset $[k] } }'\n",
       "\t Example: mlr --from f.dat put '...; unset @sums'\n",
       "\t Example: mlr --from f.dat put '...; unset @sums[\"green\"]'\n",
       "\t Example: mlr --from f.dat put '...; unset @*'\n",
       "\n",
       "   var\n",
       "       var: declares an untyped local variable in the current curly-braced scope.\n",
       "       Examples: 'var a=1', 'var xyz=\"\"'\n",
       "\n",
       "   while\n",
       "       while: introduces a while loop, or with \"do\", introduces a do-while loop.\n",
       "       The body statements must be wrapped in curly braces.\n",
       "\n",
       "   ENV\n",
       "       ENV: access to environment variables by name, e.g. '$home = ENV[\"HOME\"]'\n",
       "\n",
       "   FILENAME\n",
       "       FILENAME: evaluates to the name of the current file being processed.\n",
       "\n",
       "   FILENUM\n",
       "       FILENUM: evaluates to the number of the current file being processed,\n",
       "       starting with 1.\n",
       "\n",
       "   FNR\n",
       "       FNR: evaluates to the number of the current record within the current file\n",
       "       being processed, starting with 1. Resets at the start of each file.\n",
       "\n",
       "   IFS\n",
       "       IFS: evaluates to the input field separator from the command line.\n",
       "\n",
       "   IPS\n",
       "       IPS: evaluates to the input pair separator from the command line.\n",
       "\n",
       "   IRS\n",
       "       IRS: evaluates to the input record separator from the command line,\n",
       "       or to LF or CRLF from the input data if in autodetect mode (which is\n",
       "       the default).\n",
       "\n",
       "   M_E\n",
       "       M_E: the mathematical constant e.\n",
       "\n",
       "   M_PI\n",
       "       M_PI: the mathematical constant pi.\n",
       "\n",
       "   NF\n",
       "       NF: evaluates to the number of fields in the current record.\n",
       "\n",
       "   NR\n",
       "       NR: evaluates to the number of the current record over all files\n",
       "       being processed, starting with 1. Does not reset at the start of each file.\n",
       "\n",
       "   OFS\n",
       "       OFS: evaluates to the output field separator from the command line.\n",
       "\n",
       "   OPS\n",
       "       OPS: evaluates to the output pair separator from the command line.\n",
       "\n",
       "   ORS\n",
       "       ORS: evaluates to the output record separator from the command line,\n",
       "       or to LF or CRLF from the input data if in autodetect mode (which is\n",
       "       the default).\n",
       "\n",
       "AUTHOR\n",
       "       Miller is written by John Kerl <kerl.john.r@gmail.com>.\n",
       "\n",
       "       This manual page has been composed from Miller's help output by Eric\n",
       "       MSP Veith <eveith@veith-m.de>.\n",
       "\n",
       "SEE ALSO\n",
       "       awk(1), sed(1), cut(1), join(1), sort(1), RFC 4180: Common Format and\n",
       "       MIME Type for Comma-Separated Values (CSV) Files, the miller website\n",
       "       http://johnkerl.org/miller/doc\n",
       "\n",
       "\n",
       "\n",
       "\t\t\t\t  2020-03-17\t\t\t     MILLER(1)\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "man mlr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
